{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once I received the k-mers files from Dmitrii, here I want to explore the following points:\n",
    "- Read those files: are under processed.zip, hence we have to decompressed them first\n",
    "- Explore the data: how many k-mers I have per each species -> why only those?\n",
    "- Create a new match function between two scaffolds: how many k-mers are shared between them?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "# Import libraries\n",
    "import pandas as pd\n",
    "import feather\n",
    "import zipfile\n",
    "from sympy import isprime\n",
    "import os\n",
    "\n",
    "import k_mers_metric as kmm\n",
    "import read_k_mers as rkm\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read k-mers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"# Read feather file of each species\\n# Only the relaxed_locus file is needed\\n# Only for k prime\\n\\n# Path to the folder with the processed files\\npath = '../Data/Intermediate/k_mers/processed/'\\n\\ndf_list = []\\n\\nfor i in range(1, 101):\\n    # Check if i is prime\\n    if isprime(i):\\n        # Read files of all the folders (species)\\n        for folder in os.listdir(path):\\n            try:\\n                # Read feather file\\n                df = pd.read_feather(path + '/' + folder + '/' + str(i) + '.relaxed_locus.feather')\\n\\n                # Add species column and k column\\n                df['species'] = folder\\n                df['k'] = i\\n\\n                # Concatenate all the dataframes\\n                df_list = pd.concat([df_list, df], ignore_index=True)        \\n\\n                print('File ' + str(i) + ' of ' + folder + ' read')\\n                    #logging.info('File ' + str(i) + ' of ' + folder + ' read')\\n            except:\\n               pass\""
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''# Read feather file of each species\n",
    "# Only the relaxed_locus file is needed\n",
    "# Only for k prime\n",
    "\n",
    "# Path to the folder with the processed files\n",
    "path = '../Data/Intermediate/k_mers/processed/'\n",
    "\n",
    "df_list = []\n",
    "\n",
    "for i in range(1, 101):\n",
    "    # Check if i is prime\n",
    "    if isprime(i):\n",
    "        # Read files of all the folders (species)\n",
    "        for folder in os.listdir(path):\n",
    "            try:\n",
    "                # Read feather file\n",
    "                df = pd.read_feather(path + '/' + folder + '/' + str(i) + '.relaxed_locus.feather')\n",
    "\n",
    "                # Add species column and k column\n",
    "                df['species'] = folder\n",
    "                df['k'] = i\n",
    "\n",
    "                # Concatenate all the dataframes\n",
    "                df_list = pd.concat([df_list, df], ignore_index=True)        \n",
    "\n",
    "                print('File ' + str(i) + ' of ' + folder + ' read')\n",
    "                    #logging.info('File ' + str(i) + ' of ' + folder + ' read')\n",
    "            except:\n",
    "               pass'''\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"import glob\\nimport pandas as pd\\nfrom sympy import isprime\\nimport multiprocessing\\n\\n# Path to the folder with the processed files\\npath = '../Data/Intermediate/k_mers/processed/'\\n\\n# Define a function to read a single feather file and return a dictionary\\ndef read_file(filepath):\\n    try:\\n        df = pd.read_feather(filepath)\\n        folder, filename = filepath.split('/')[-2:]\\n        k = int(filename.split('.')[0])\\n        return {'species': folder, 'k': k, 'data': df}\\n    except:\\n        return None\\n\\n# Use glob to find all the feather files\\nfilepaths = glob.glob(path + '/**/*.relaxed_locus.feather', recursive=True)\\n\\n# Use multiprocessing to read the files in parallel\\nwith multiprocessing.Pool() as pool:\\n    data_dicts = pool.map(read_file, filepaths)\\n\\n# Filter out any failed reads\\ndata_dicts = [d for d in data_dicts if d is not None]\\n\\n# Concatenate the dictionaries into a single DataFrame\\n#df = pd.concat([d['data'].assign(species=d['species'], k=d['k']) for d in data_dicts], ignore_index=True)\\n\\nprint('Finished reading', len(data_dicts), 'rows')\""
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''import glob\n",
    "import pandas as pd\n",
    "from sympy import isprime\n",
    "import multiprocessing\n",
    "\n",
    "# Path to the folder with the processed files\n",
    "path = '../Data/Intermediate/k_mers/processed/'\n",
    "\n",
    "# Define a function to read a single feather file and return a dictionary\n",
    "def read_file(filepath):\n",
    "    try:\n",
    "        df = pd.read_feather(filepath)\n",
    "        folder, filename = filepath.split('/')[-2:]\n",
    "        k = int(filename.split('.')[0])\n",
    "        return {'species': folder, 'k': k, 'data': df}\n",
    "    except:\n",
    "        return None\n",
    "\n",
    "# Use glob to find all the feather files\n",
    "filepaths = glob.glob(path + '/**/*.relaxed_locus.feather', recursive=True)\n",
    "\n",
    "# Use multiprocessing to read the files in parallel\n",
    "with multiprocessing.Pool() as pool:\n",
    "    data_dicts = pool.map(read_file, filepaths)\n",
    "\n",
    "# Filter out any failed reads\n",
    "data_dicts = [d for d in data_dicts if d is not None]\n",
    "\n",
    "# Concatenate the dictionaries into a single DataFrame\n",
    "#df = pd.concat([d['data'].assign(species=d['species'], k=d['k']) for d in data_dicts], ignore_index=True)\n",
    "\n",
    "print('Finished reading', len(data_dicts), 'rows')'''\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The good one\n",
    "\n",
    "import os\n",
    "import pandas as pd\n",
    "from sympy import isprime\n",
    "\n",
    "# Path to the folder with the processed files\n",
    "path = '../Data/Intermediate/k_mers/processed/'\n",
    "\n",
    "# Define a function to read a single feather file and return a dictionary\n",
    "def read_file(filepath, mode = 'strict'):\n",
    "    try:\n",
    "        if mode in filepath:\n",
    "            df = pd.read_feather(filepath)\n",
    "            folder, filename = os.path.split(filepath)\n",
    "            k = int(filename.split('.')[0])\n",
    "            species = folder.split('/')[-1]\n",
    "            \n",
    "            print('File ' + str(k) + ' of ' + species + ' read')\n",
    "            #logging.info('File ' + str(i) + ' of ' + folder + ' read')\n",
    "\n",
    "            return {'species_gcf_id': species, 'k': k, 'data': df}\n",
    "    except:\n",
    "        return None\n",
    "\n",
    "def process_folder(path):\n",
    "    # Read all the feather files and store the data in a list of dictionaries\n",
    "    data_list = []\n",
    "    for folder in os.listdir(path):\n",
    "        folder_path = os.path.join(path, folder)\n",
    "        print(folder_path)\n",
    "        for filename in os.listdir(folder_path):\n",
    "            filepath = os.path.join(folder_path, filename)\n",
    "            k = filename.split('.')[0]\n",
    "            # Check if it is a number and if it is prime\n",
    "            if k.isdigit():\n",
    "                if isprime(int(k)):\n",
    "                    data_dict = read_file(filepath, mode = 'strict')\n",
    "                    if data_dict is not None:\n",
    "                        data_list.append(data_dict)\n",
    "    return data_list\n",
    "\n",
    "data_list = rkm.process_folder(path, mode='strict')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'data'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[9], line 2\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[39m# Convert the list of dictionaries into a single DataFrame, including the species and k columns\u001b[39;00m\n\u001b[1;32m----> 2\u001b[0m df_list \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mconcat([d[\u001b[39m'\u001b[39m\u001b[39mdata\u001b[39m\u001b[39m'\u001b[39m]\u001b[39m.\u001b[39massign(species_gcf_id\u001b[39m=\u001b[39md[\u001b[39m'\u001b[39m\u001b[39mspecies_gcf_id\u001b[39m\u001b[39m'\u001b[39m], k\u001b[39m=\u001b[39md[\u001b[39m'\u001b[39m\u001b[39mk\u001b[39m\u001b[39m'\u001b[39m]) \u001b[39mfor\u001b[39;00m d \u001b[39min\u001b[39;00m data_list], ignore_index\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m)\n\u001b[0;32m      3\u001b[0m \u001b[39m#df_list = pd.concat([d['data'] for d in data_list], ignore_index=True)\u001b[39;00m\n\u001b[0;32m      4\u001b[0m df_list \u001b[39m=\u001b[39m df_list[[\u001b[39m'\u001b[39m\u001b[39mspecies_gcf_id\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mk\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mk_mer\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mreplicon_accession\u001b[39m\u001b[39m'\u001b[39m]]\u001b[39m.\u001b[39mrename(columns\u001b[39m=\u001b[39m{\u001b[39m'\u001b[39m\u001b[39mreplicon_accession\u001b[39m\u001b[39m'\u001b[39m: \u001b[39m'\u001b[39m\u001b[39mscaffold\u001b[39m\u001b[39m'\u001b[39m})\n",
      "Cell \u001b[1;32mIn[9], line 2\u001b[0m, in \u001b[0;36m<listcomp>\u001b[1;34m(.0)\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[39m# Convert the list of dictionaries into a single DataFrame, including the species and k columns\u001b[39;00m\n\u001b[1;32m----> 2\u001b[0m df_list \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mconcat([d[\u001b[39m'\u001b[39;49m\u001b[39mdata\u001b[39;49m\u001b[39m'\u001b[39;49m]\u001b[39m.\u001b[39massign(species_gcf_id\u001b[39m=\u001b[39md[\u001b[39m'\u001b[39m\u001b[39mspecies_gcf_id\u001b[39m\u001b[39m'\u001b[39m], k\u001b[39m=\u001b[39md[\u001b[39m'\u001b[39m\u001b[39mk\u001b[39m\u001b[39m'\u001b[39m]) \u001b[39mfor\u001b[39;00m d \u001b[39min\u001b[39;00m data_list], ignore_index\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m)\n\u001b[0;32m      3\u001b[0m \u001b[39m#df_list = pd.concat([d['data'] for d in data_list], ignore_index=True)\u001b[39;00m\n\u001b[0;32m      4\u001b[0m df_list \u001b[39m=\u001b[39m df_list[[\u001b[39m'\u001b[39m\u001b[39mspecies_gcf_id\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mk\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mk_mer\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mreplicon_accession\u001b[39m\u001b[39m'\u001b[39m]]\u001b[39m.\u001b[39mrename(columns\u001b[39m=\u001b[39m{\u001b[39m'\u001b[39m\u001b[39mreplicon_accession\u001b[39m\u001b[39m'\u001b[39m: \u001b[39m'\u001b[39m\u001b[39mscaffold\u001b[39m\u001b[39m'\u001b[39m})\n",
      "\u001b[1;31mKeyError\u001b[0m: 'data'"
     ]
    }
   ],
   "source": [
    "# Convert the list of dictionaries into a single DataFrame, including the species and k columns\n",
    "df_list = pd.concat([d['data'].assign(species_gcf_id=d['species_gcf_id'], k=d['k']) for d in data_list], ignore_index=True)\n",
    "#df_list = pd.concat([d['data'] for d in data_list], ignore_index=True)\n",
    "df_list = df_list[['species_gcf_id', 'k', 'k_mer', 'replicon_accession']].rename(columns={'replicon_accession': 'scaffold'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Sanity check\n",
    "df_list['count_k_mers'] = df_list['k_mer'].str.count('_') + 1\n",
    "df_list.loc[df_list['count_k_mers'].astype(int) != df_list['k'].astype(int), 'species'].unique()'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "test = pd.read_feather('../Data/Intermediate/k_mers/processed/GCF_004010195.1_ASM401019v2_genomic/2.strict_locus.feather')\n",
    "test['count_k_mers'] = test['k_mer'].str.count('_') + 1\n",
    "test.loc[test['count_k_mers'] != 2]\n",
    "df_list.loc[df_list['count_k_mers'].astype(int) != df_list['k'].astype(int), ['count_k_mers', 'species']].value_counts()'''"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Species name\n",
    "Correct the species name by searching them the readable ones in the config.yaml file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import yaml\n",
    "\n",
    "def map_species_to_file():\n",
    "    # Open config.yaml file\n",
    "    with open('config.yaml', 'r') as f:\n",
    "        config = yaml.load(f, Loader=yaml.FullLoader)\n",
    "\n",
    "    # Create a dictionary with the filename and the species name from the config file\n",
    "    # First, take only the organisms with a filename\n",
    "    organisms = {k: v for k, v in config['organisms'].items() if v['filename'] is not None}\n",
    "\n",
    "    # Create the dictionary with the filename as key and the species as value\n",
    "    file_species = {v['filename']: v['name'] for k, v in organisms.items()}\n",
    "\n",
    "    return file_species\n",
    "\n",
    "file_species = map_species_to_file()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a column in the dataframe with the species name\n",
    "df_list['species'] = df_list['species_gcf_id'].map(file_species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_list.to_feather('../Data/Intermediate/k_mers/processed/df_k_mers.feather')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# To read them\n",
    "df_k_mers = pd.read_feather('../Data/Intermediate/k_mers/processed/df_k_mers.feather')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Algorithm\n",
    "\n",
    "Aim: given a specific scaffold, know how many k-mers are shared with the rest of scaffolds."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\bertr\\AppData\\Local\\Temp\\ipykernel_31248\\2548606989.py:4: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.\n",
      "  df_k_mers_species = df_k_mers_scaffold.groupby(['species', 'k'], as_index=False).sum()\n"
     ]
    }
   ],
   "source": [
    "def count_k_mers(df_k_mers, output_folder= '../Data/Intermediate/k_mers/'):\n",
    "    # From df_k_mers, create a dataframe with the number of k_mers per scaffold / species\n",
    "    df_k_mers_scaffold = df_k_mers.groupby(['species', 'scaffold', 'k'], as_index=False).nunique()\n",
    "    df_k_mers_species = df_k_mers_scaffold.groupby(['species', 'k'], as_index=False).sum()\n",
    "\n",
    "    df_k_mers_scaffold = df_k_mers_scaffold[['species', 'scaffold', 'k', 'k_mer']].rename(columns = {'k_mer': 'total_k_mers'})\n",
    "    df_k_mers_species = df_k_mers_species[['species', 'k', 'k_mer']].rename(columns = {'k_mer': 'total_k_mers'})\n",
    "\n",
    "    # Save them into a feather file\n",
    "    df_k_mers_scaffold.to_feather(output_folder + 'count_k_mers_scaffold.feather')\n",
    "    df_k_mers_species.to_feather(output_folder + 'count_k_mers_species.feather')\n",
    "\n",
    "count_k_mers(df_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def shared_k_mers(origin, level, df_k_mers):\n",
    "\n",
    "    # Get the k-mers of the scaffold or species\n",
    "    df_origin = df_k_mers[df_k_mers[level] == origin]\n",
    "\n",
    "    # Get the k-mers shared with the rest of scaffolds or species\n",
    "    df_shared = df_origin.merge(df_k_mers, on = ['k', 'k_mer'])\n",
    "\n",
    "    df_shared.rename(columns = {\n",
    "                        'species_y': 'species',\n",
    "                        'scaffold_y': 'scaffold'}, inplace = True)\n",
    "\n",
    "    # Remove the k-mers shared within the same species\n",
    "    df_shared = df_shared.loc[df_shared['species_x'] != df_shared['species'], ['k', level, 'k_mer']]\n",
    "\n",
    "    # Get the number of k-mers shared with the rest of scaffolds\n",
    "    df_common_k_mers = df_shared.groupby(['k', level]).count().reset_index()\n",
    "    df_common_k_mers = df_common_k_mers[['k', level, 'k_mer']].rename(columns = {'k_mer': 'common_k_mers'})\n",
    "\n",
    "    return df_common_k_mers\n",
    "    \n",
    "def generated_k_mers(origin, level, k_mers_counter):\n",
    "    # Let's remove the rows with NaN values\n",
    "    # But then, we will remove the information from the scaffold of interest\n",
    "    # Let's keep it in a separate dataframe\n",
    "    k_mers_counter_origin= k_mers_counter[k_mers_counter[level] == origin]\n",
    "\n",
    "    k_mers_counter = k_mers_counter.loc[k_mers_counter['common_k_mers'].notnull()]\n",
    "\n",
    "    # Merge the two dataframes, creating a new column containing the number of k-mers generated by the original scaffold\n",
    "    k_mers_counter = k_mers_counter.merge(k_mers_counter_origin[['k', 'total_k_mers', 'length']], on = ['k'], how = 'left', suffixes = ('', '_origin'))\n",
    "\n",
    "    # Compute the number of k-mers generated by the two scaffolds\n",
    "    k_mers_counter['generated_k_mers'] = k_mers_counter['total_k_mers'] + k_mers_counter['total_k_mers_origin'] - k_mers_counter['common_k_mers']\n",
    "\n",
    "    return k_mers_counter\n",
    "\n",
    "# Metric computation\n",
    "def metric_computation(k_mers_counter):\n",
    "    # Let's keep only with the rows with all the information\n",
    "    k_mers_counter = k_mers_counter.dropna()\n",
    "\n",
    "    # Compute the probability of finding a common k-mer\n",
    "    k_mers_counter['probability'] = k_mers_counter['common_k_mers'] / k_mers_counter['generated_k_mers']\n",
    "\n",
    "    # First, we want to apply the desired formula:\n",
    "    # (1 - probability) * (1 - (m / (n*k^2)))\n",
    "    # Where: m, n are the length of the two scaffolds/species with m being the shortest represented in length and length_origin columns\n",
    "    k_mers_counter['metric'] = (1 - k_mers_counter['probability']) * (1 - (k_mers_counter[['length', 'length_origin']].apply(min, axis=1) / (k_mers_counter[['length', 'length_origin']].apply(max, axis=1) * k_mers_counter['k']**2)))\n",
    "\n",
    "    # Leave the columnns of interest, so we can aggregate the metric and sort it by species in any cases\n",
    "    df_metric = k_mers_counter.drop(['total_k_mers', 'total_k_mers_origin', 'probability', 'common_k_mers', 'generated_k_mers', 'length_origin', 'length'], axis = 1)\n",
    "\n",
    "    # Do the product of the metric for each scaffold/species among all the k values\n",
    "    if level == 'scaffold':\n",
    "        return (df_metric\n",
    "                .groupby(['species', 'scaffold'], as_index=False)\n",
    "                .prod()\n",
    "                .sort_values(by = ['species','metric'], ascending = False))\n",
    "    elif level == 'species':\n",
    "        return (df_metric\n",
    "                .groupby('species', as_index=False)\n",
    "                .prod()\n",
    "                .sort_values(by = ['species', 'metric'], ascending = False))\n",
    "\n",
    "def k_mers_to_metric(origin, level, k_mers_counter, df_length, df_k_mers):\n",
    "    \n",
    "    # Adapt the information of the scaffold length to the species level\n",
    "    #if level == 'species':\n",
    "    df_length = df_length.groupby([level], as_index=False).sum()\n",
    "\n",
    "    # Add the length of the scaffold or species to the dataframe\n",
    "    k_mers_counter = k_mers_counter.merge(df_length[[level, 'length']], on = level, how = 'left')\n",
    "\n",
    "    # Now, we need to compute the number of k-mers shared from the origin with the rest of elements\n",
    "    print('Computing shared k-mers for: ' + origin + ' '  + level)\n",
    "    df_common_k_mers = shared_k_mers(origin, level, df_k_mers)\n",
    "\n",
    "    k_mers_counter = k_mers_counter.join(df_common_k_mers.set_index(['k', level]), on=['k', level], how='left')\n",
    "\n",
    "    # Now, we have the total k-mers per scaffols and the number of k-mers shared with the rest of scaffolds, i.e., the intersection between the two groups\n",
    "    # We want to compute the number of k-mers generated by the two scaffolds\n",
    "    # We can do this by computing the number of k-mers generated by the two scaffolds and subtracting the number of k-mers shared with the rest of scaffolds\n",
    "    # Done under generated_k_mers function\n",
    "\n",
    "    k_mers_counter = generated_k_mers(origin, level, k_mers_counter)\n",
    "\n",
    "    # Now, we can compute the metric\n",
    "    return metric_computation(k_mers_counter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['k', 'k_mer', 'scaffold', 'species_gcf_id'], dtype='object')"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_list.columns.difference(['species'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\bertr\\LRZ Sync+Share\\Thesis\\Code\\k_mers_metric.py:92: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.\n",
      "  df_length = df_length.groupby([level], as_index=False).sum()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Computing shared k-mers for: NC_000001.11 scaffold\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\bertr\\LRZ Sync+Share\\Thesis\\Code\\k_mers_metric.py:67: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  k_mers_counter['probability'] = k_mers_counter['common_k_mers'] / k_mers_counter['generated_k_mers']\n",
      "c:\\Users\\bertr\\LRZ Sync+Share\\Thesis\\Code\\k_mers_metric.py:72: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  k_mers_counter['metric'] = (1 - k_mers_counter['probability']) * (1 - (k_mers_counter[['length', 'length_origin']].apply(min, axis=1) / (k_mers_counter[['length', 'length_origin']].apply(max, axis=1) * k_mers_counter['k']**2)))\n"
     ]
    }
   ],
   "source": [
    "import k_mers_metric as kmm\n",
    "\n",
    "# Given a specific scaffold, know how many k-mers shared with the rest of scaffolds\n",
    "origin = 'NC_000001.11'\n",
    "level = 'scaffold'\n",
    "\n",
    "k_mers_counter = pd.read_feather('../Data/Intermediate/k_mers/count_k_mers_' + level + '.feather')\n",
    "\n",
    "# For computing the similarity metric between two scaffolds or species, we need:\n",
    "# - The probability of finding a common k-mer: coomon k-mers / generated k-mers between two scafffolds or species\n",
    "# - Length of each scaffold\n",
    "# - k of k-mer\n",
    "\n",
    "# First, let us add its length information into the k-mers_counter dataframe\n",
    "# This is already computed under df_scaffold_length file\n",
    "try:\n",
    "    df_length = pd.read_csv('../Data/Intermediate/df_' + level + '_length.csv')\n",
    "    df_length = (df_length[['Specie', 'Replicon Accession', 'Number of genes']]\n",
    "                        .rename(columns = {'Specie': 'species', 'Replicon Accession': 'scaffold', 'Number of genes': 'length'}))\n",
    "except:\n",
    "    print('Please, check the columns of the file df_scaffold_length.csv')\n",
    "\n",
    "scaffolds_order = kmm.k_mers_to_metric(origin, level, k_mers_counter, df_length, df_list)    #, 'metric']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>scaffold</th>\n",
       "      <th>k</th>\n",
       "      <th>metric</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054382.1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.916613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054376.1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.869437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054388.1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.868704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054380.1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.861646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054384.1</td>\n",
       "      <td>30</td>\n",
       "      <td>0.857382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054387.1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.827235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054381.1</td>\n",
       "      <td>30</td>\n",
       "      <td>0.823000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054386.1</td>\n",
       "      <td>30</td>\n",
       "      <td>0.817991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054383.1</td>\n",
       "      <td>210</td>\n",
       "      <td>0.799004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054371.1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.798263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054372.1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.792320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054378.1</td>\n",
       "      <td>30</td>\n",
       "      <td>0.788809</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054379.1</td>\n",
       "      <td>210</td>\n",
       "      <td>0.771229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054375.1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.768872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054374.1</td>\n",
       "      <td>30</td>\n",
       "      <td>0.762977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054373.1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.758905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054385.1</td>\n",
       "      <td>30</td>\n",
       "      <td>0.749960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054377.1</td>\n",
       "      <td>30</td>\n",
       "      <td>0.731555</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                species     scaffold    k    metric\n",
       "11  African clawed frog  NC_054382.1    2  0.916613\n",
       "5   African clawed frog  NC_054376.1    2  0.869437\n",
       "17  African clawed frog  NC_054388.1    2  0.868704\n",
       "9   African clawed frog  NC_054380.1    6  0.861646\n",
       "13  African clawed frog  NC_054384.1   30  0.857382\n",
       "16  African clawed frog  NC_054387.1    2  0.827235\n",
       "10  African clawed frog  NC_054381.1   30  0.823000\n",
       "15  African clawed frog  NC_054386.1   30  0.817991\n",
       "12  African clawed frog  NC_054383.1  210  0.799004\n",
       "0   African clawed frog  NC_054371.1    2  0.798263\n",
       "1   African clawed frog  NC_054372.1    6  0.792320\n",
       "7   African clawed frog  NC_054378.1   30  0.788809\n",
       "8   African clawed frog  NC_054379.1  210  0.771229\n",
       "4   African clawed frog  NC_054375.1    6  0.768872\n",
       "3   African clawed frog  NC_054374.1   30  0.762977\n",
       "2   African clawed frog  NC_054373.1    6  0.758905\n",
       "14  African clawed frog  NC_054385.1   30  0.749960\n",
       "6   African clawed frog  NC_054377.1   30  0.731555"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scaffolds_order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species_name</th>\n",
       "      <th>replicon_accession</th>\n",
       "      <th>metric</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054382.1</td>\n",
       "      <td>0.916613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054376.1</td>\n",
       "      <td>0.869437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054388.1</td>\n",
       "      <td>0.868704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054380.1</td>\n",
       "      <td>0.861646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054384.1</td>\n",
       "      <td>0.857382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054387.1</td>\n",
       "      <td>0.827235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054381.1</td>\n",
       "      <td>0.823000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054386.1</td>\n",
       "      <td>0.817991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054383.1</td>\n",
       "      <td>0.799004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054371.1</td>\n",
       "      <td>0.798263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054372.1</td>\n",
       "      <td>0.792320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054378.1</td>\n",
       "      <td>0.788809</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054379.1</td>\n",
       "      <td>0.771229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054375.1</td>\n",
       "      <td>0.768872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054374.1</td>\n",
       "      <td>0.762977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054373.1</td>\n",
       "      <td>0.758905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054385.1</td>\n",
       "      <td>0.749960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054377.1</td>\n",
       "      <td>0.731555</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           species_name replicon_accession    metric\n",
       "11  African clawed frog        NC_054382.1  0.916613\n",
       "5   African clawed frog        NC_054376.1  0.869437\n",
       "17  African clawed frog        NC_054388.1  0.868704\n",
       "9   African clawed frog        NC_054380.1  0.861646\n",
       "13  African clawed frog        NC_054384.1  0.857382\n",
       "16  African clawed frog        NC_054387.1  0.827235\n",
       "10  African clawed frog        NC_054381.1  0.823000\n",
       "15  African clawed frog        NC_054386.1  0.817991\n",
       "12  African clawed frog        NC_054383.1  0.799004\n",
       "0   African clawed frog        NC_054371.1  0.798263\n",
       "1   African clawed frog        NC_054372.1  0.792320\n",
       "7   African clawed frog        NC_054378.1  0.788809\n",
       "8   African clawed frog        NC_054379.1  0.771229\n",
       "4   African clawed frog        NC_054375.1  0.768872\n",
       "3   African clawed frog        NC_054374.1  0.762977\n",
       "2   African clawed frog        NC_054373.1  0.758905\n",
       "14  African clawed frog        NC_054385.1  0.749960\n",
       "6   African clawed frog        NC_054377.1  0.731555"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metric"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species_name</th>\n",
       "      <th>replicon_accession</th>\n",
       "      <th>k</th>\n",
       "      <th>total_k_mers</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054371.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1766</td>\n",
       "      <td>1848.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054371.1</td>\n",
       "      <td>3</td>\n",
       "      <td>1778</td>\n",
       "      <td>1848.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054371.1</td>\n",
       "      <td>5</td>\n",
       "      <td>1778</td>\n",
       "      <td>1848.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054371.1</td>\n",
       "      <td>7</td>\n",
       "      <td>1776</td>\n",
       "      <td>1848.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054371.1</td>\n",
       "      <td>11</td>\n",
       "      <td>1772</td>\n",
       "      <td>1848.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5386</th>\n",
       "      <td>Human</td>\n",
       "      <td>NT_187693.1</td>\n",
       "      <td>13</td>\n",
       "      <td>17</td>\n",
       "      <td>45.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5387</th>\n",
       "      <td>Human</td>\n",
       "      <td>NT_187693.1</td>\n",
       "      <td>17</td>\n",
       "      <td>13</td>\n",
       "      <td>45.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5388</th>\n",
       "      <td>Human</td>\n",
       "      <td>NT_187693.1</td>\n",
       "      <td>19</td>\n",
       "      <td>11</td>\n",
       "      <td>45.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5389</th>\n",
       "      <td>Human</td>\n",
       "      <td>NT_187693.1</td>\n",
       "      <td>23</td>\n",
       "      <td>7</td>\n",
       "      <td>45.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5390</th>\n",
       "      <td>Human</td>\n",
       "      <td>NT_187693.1</td>\n",
       "      <td>29</td>\n",
       "      <td>1</td>\n",
       "      <td>45.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1430 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             species_name replicon_accession   k  total_k_mers  length\n",
       "46    African clawed frog        NC_054371.1   2          1766  1848.0\n",
       "47    African clawed frog        NC_054371.1   3          1778  1848.0\n",
       "48    African clawed frog        NC_054371.1   5          1778  1848.0\n",
       "49    African clawed frog        NC_054371.1   7          1776  1848.0\n",
       "50    African clawed frog        NC_054371.1  11          1772  1848.0\n",
       "...                   ...                ...  ..           ...     ...\n",
       "5386                Human        NT_187693.1  13            17    45.0\n",
       "5387                Human        NT_187693.1  17            13    45.0\n",
       "5388                Human        NT_187693.1  19            11    45.0\n",
       "5389                Human        NT_187693.1  23             7    45.0\n",
       "5390                Human        NT_187693.1  29             1    45.0\n",
       "\n",
       "[1430 rows x 5 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k_mers_counter[k_mers_counter['length'].notnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if k_mers_counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species_name</th>\n",
       "      <th>replicon_accession</th>\n",
       "      <th>k</th>\n",
       "      <th>total_k_mers</th>\n",
       "      <th>length</th>\n",
       "      <th>common_k_mers</th>\n",
       "      <th>total_k_mers_origin</th>\n",
       "      <th>generated_k_mers</th>\n",
       "      <th>probability</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aardvark</td>\n",
       "      <td>NW_006921</td>\n",
       "      <td>2</td>\n",
       "      <td>11830</td>\n",
       "      <td>NaN</td>\n",
       "      <td>163.0</td>\n",
       "      <td>496</td>\n",
       "      <td>12163.0</td>\n",
       "      <td>0.013401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Aardvark</td>\n",
       "      <td>NW_006921</td>\n",
       "      <td>3</td>\n",
       "      <td>11702</td>\n",
       "      <td>NaN</td>\n",
       "      <td>102.0</td>\n",
       "      <td>508</td>\n",
       "      <td>12108.0</td>\n",
       "      <td>0.008424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Aardvark</td>\n",
       "      <td>NW_006921</td>\n",
       "      <td>5</td>\n",
       "      <td>10970</td>\n",
       "      <td>NaN</td>\n",
       "      <td>74.0</td>\n",
       "      <td>516</td>\n",
       "      <td>11412.0</td>\n",
       "      <td>0.006484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Aardvark</td>\n",
       "      <td>NW_006921</td>\n",
       "      <td>7</td>\n",
       "      <td>10266</td>\n",
       "      <td>NaN</td>\n",
       "      <td>62.0</td>\n",
       "      <td>520</td>\n",
       "      <td>10724.0</td>\n",
       "      <td>0.005781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Aardvark</td>\n",
       "      <td>NW_006921</td>\n",
       "      <td>11</td>\n",
       "      <td>9018</td>\n",
       "      <td>NaN</td>\n",
       "      <td>43.0</td>\n",
       "      <td>520</td>\n",
       "      <td>9495.0</td>\n",
       "      <td>0.004529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>648</th>\n",
       "      <td>Zebra Finch</td>\n",
       "      <td>NC_044234</td>\n",
       "      <td>2</td>\n",
       "      <td>159</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>496</td>\n",
       "      <td>651.0</td>\n",
       "      <td>0.006144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>649</th>\n",
       "      <td>Zebra Finch</td>\n",
       "      <td>NC_044234</td>\n",
       "      <td>3</td>\n",
       "      <td>158</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>508</td>\n",
       "      <td>663.0</td>\n",
       "      <td>0.004525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>650</th>\n",
       "      <td>Zebra Finch</td>\n",
       "      <td>NC_044234</td>\n",
       "      <td>5</td>\n",
       "      <td>156</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>516</td>\n",
       "      <td>671.0</td>\n",
       "      <td>0.001490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>651</th>\n",
       "      <td>Zebra Finch</td>\n",
       "      <td>NC_044236</td>\n",
       "      <td>2</td>\n",
       "      <td>144</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>496</td>\n",
       "      <td>639.0</td>\n",
       "      <td>0.001565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>652</th>\n",
       "      <td>Zebra Finch</td>\n",
       "      <td>NC_054769</td>\n",
       "      <td>2</td>\n",
       "      <td>55</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>496</td>\n",
       "      <td>550.0</td>\n",
       "      <td>0.001818</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>653 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    species_name replicon_accession   k  total_k_mers  length  common_k_mers  \\\n",
       "0       Aardvark          NW_006921   2         11830     NaN          163.0   \n",
       "1       Aardvark          NW_006921   3         11702     NaN          102.0   \n",
       "2       Aardvark          NW_006921   5         10970     NaN           74.0   \n",
       "3       Aardvark          NW_006921   7         10266     NaN           62.0   \n",
       "4       Aardvark          NW_006921  11          9018     NaN           43.0   \n",
       "..           ...                ...  ..           ...     ...            ...   \n",
       "648  Zebra Finch          NC_044234   2           159     NaN            4.0   \n",
       "649  Zebra Finch          NC_044234   3           158     NaN            3.0   \n",
       "650  Zebra Finch          NC_044234   5           156     NaN            1.0   \n",
       "651  Zebra Finch          NC_044236   2           144     NaN            1.0   \n",
       "652  Zebra Finch          NC_054769   2            55     NaN            1.0   \n",
       "\n",
       "     total_k_mers_origin  generated_k_mers  probability  \n",
       "0                    496           12163.0     0.013401  \n",
       "1                    508           12108.0     0.008424  \n",
       "2                    516           11412.0     0.006484  \n",
       "3                    520           10724.0     0.005781  \n",
       "4                    520            9495.0     0.004529  \n",
       "..                   ...               ...          ...  \n",
       "648                  496             651.0     0.006144  \n",
       "649                  508             663.0     0.004525  \n",
       "650                  516             671.0     0.001490  \n",
       "651                  496             639.0     0.001565  \n",
       "652                  496             550.0     0.001818  \n",
       "\n",
       "[653 rows x 9 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k_mers_counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Specie</th>\n",
       "      <th>#Replicon Name</th>\n",
       "      <th>Replicon Accession</th>\n",
       "      <th>Number of genes</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aadvark</td>\n",
       "      <td>MT</td>\n",
       "      <td>NC_002078.1</td>\n",
       "      <td>13</td>\n",
       "      <td>2744</td>\n",
       "      <td>15310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Aadvark</td>\n",
       "      <td>Un</td>\n",
       "      <td>NW_006921588.1</td>\n",
       "      <td>129</td>\n",
       "      <td>2349339</td>\n",
       "      <td>39525097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Aadvark</td>\n",
       "      <td>Un</td>\n",
       "      <td>NW_006921589.1</td>\n",
       "      <td>169</td>\n",
       "      <td>30547</td>\n",
       "      <td>38336954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Aadvark</td>\n",
       "      <td>Un</td>\n",
       "      <td>NW_006921590.1</td>\n",
       "      <td>77</td>\n",
       "      <td>719496</td>\n",
       "      <td>36587654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Aadvark</td>\n",
       "      <td>Un</td>\n",
       "      <td>NW_006921591.1</td>\n",
       "      <td>33</td>\n",
       "      <td>8315</td>\n",
       "      <td>35191965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37053</th>\n",
       "      <td>Xtropicalis</td>\n",
       "      <td>Un</td>\n",
       "      <td>NW_016689751.1</td>\n",
       "      <td>1</td>\n",
       "      <td>557</td>\n",
       "      <td>1575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37054</th>\n",
       "      <td>Xtropicalis</td>\n",
       "      <td>Un</td>\n",
       "      <td>NW_016689781.1</td>\n",
       "      <td>1</td>\n",
       "      <td>1617</td>\n",
       "      <td>2094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37055</th>\n",
       "      <td>Xtropicalis</td>\n",
       "      <td>Un</td>\n",
       "      <td>NW_016689786.1</td>\n",
       "      <td>1</td>\n",
       "      <td>888</td>\n",
       "      <td>1010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37056</th>\n",
       "      <td>Xtropicalis</td>\n",
       "      <td>Un</td>\n",
       "      <td>NW_016689912.1</td>\n",
       "      <td>1</td>\n",
       "      <td>326</td>\n",
       "      <td>410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37057</th>\n",
       "      <td>Xtropicalis</td>\n",
       "      <td>Un</td>\n",
       "      <td>NW_016690069.1</td>\n",
       "      <td>1</td>\n",
       "      <td>760</td>\n",
       "      <td>990</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>37058 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Specie #Replicon Name Replicon Accession  Number of genes  \\\n",
       "0          Aadvark             MT        NC_002078.1               13   \n",
       "1          Aadvark             Un     NW_006921588.1              129   \n",
       "2          Aadvark             Un     NW_006921589.1              169   \n",
       "3          Aadvark             Un     NW_006921590.1               77   \n",
       "4          Aadvark             Un     NW_006921591.1               33   \n",
       "...            ...            ...                ...              ...   \n",
       "37053  Xtropicalis             Un     NW_016689751.1                1   \n",
       "37054  Xtropicalis             Un     NW_016689781.1                1   \n",
       "37055  Xtropicalis             Un     NW_016689786.1                1   \n",
       "37056  Xtropicalis             Un     NW_016689912.1                1   \n",
       "37057  Xtropicalis             Un     NW_016690069.1                1   \n",
       "\n",
       "         Start      Stop  \n",
       "0         2744     15310  \n",
       "1      2349339  39525097  \n",
       "2        30547  38336954  \n",
       "3       719496  36587654  \n",
       "4         8315  35191965  \n",
       "...        ...       ...  \n",
       "37053      557      1575  \n",
       "37054     1617      2094  \n",
       "37055      888      1010  \n",
       "37056      326       410  \n",
       "37057      760       990  \n",
       "\n",
       "[37058 rows x 6 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_scaffold_length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species_name</th>\n",
       "      <th>replicon_accession</th>\n",
       "      <th>k</th>\n",
       "      <th>total_k_mers</th>\n",
       "      <th>common_k_mers</th>\n",
       "      <th>total_k_mers_origin</th>\n",
       "      <th>generated_k_mers</th>\n",
       "      <th>probability</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aardvark</td>\n",
       "      <td>NW_006921</td>\n",
       "      <td>2</td>\n",
       "      <td>11830</td>\n",
       "      <td>163.0</td>\n",
       "      <td>496</td>\n",
       "      <td>12163.0</td>\n",
       "      <td>0.013401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Aardvark</td>\n",
       "      <td>NW_006921</td>\n",
       "      <td>3</td>\n",
       "      <td>11702</td>\n",
       "      <td>102.0</td>\n",
       "      <td>508</td>\n",
       "      <td>12108.0</td>\n",
       "      <td>0.008424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Aardvark</td>\n",
       "      <td>NW_006921</td>\n",
       "      <td>5</td>\n",
       "      <td>10970</td>\n",
       "      <td>74.0</td>\n",
       "      <td>516</td>\n",
       "      <td>11412.0</td>\n",
       "      <td>0.006484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Aardvark</td>\n",
       "      <td>NW_006921</td>\n",
       "      <td>7</td>\n",
       "      <td>10266</td>\n",
       "      <td>62.0</td>\n",
       "      <td>520</td>\n",
       "      <td>10724.0</td>\n",
       "      <td>0.005781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Aardvark</td>\n",
       "      <td>NW_006921</td>\n",
       "      <td>11</td>\n",
       "      <td>9018</td>\n",
       "      <td>43.0</td>\n",
       "      <td>520</td>\n",
       "      <td>9495.0</td>\n",
       "      <td>0.004529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>648</th>\n",
       "      <td>Zebra Finch</td>\n",
       "      <td>NC_044234</td>\n",
       "      <td>2</td>\n",
       "      <td>159</td>\n",
       "      <td>4.0</td>\n",
       "      <td>496</td>\n",
       "      <td>651.0</td>\n",
       "      <td>0.006144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>649</th>\n",
       "      <td>Zebra Finch</td>\n",
       "      <td>NC_044234</td>\n",
       "      <td>3</td>\n",
       "      <td>158</td>\n",
       "      <td>3.0</td>\n",
       "      <td>508</td>\n",
       "      <td>663.0</td>\n",
       "      <td>0.004525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>650</th>\n",
       "      <td>Zebra Finch</td>\n",
       "      <td>NC_044234</td>\n",
       "      <td>5</td>\n",
       "      <td>156</td>\n",
       "      <td>1.0</td>\n",
       "      <td>516</td>\n",
       "      <td>671.0</td>\n",
       "      <td>0.001490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>651</th>\n",
       "      <td>Zebra Finch</td>\n",
       "      <td>NC_044236</td>\n",
       "      <td>2</td>\n",
       "      <td>144</td>\n",
       "      <td>1.0</td>\n",
       "      <td>496</td>\n",
       "      <td>639.0</td>\n",
       "      <td>0.001565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>652</th>\n",
       "      <td>Zebra Finch</td>\n",
       "      <td>NC_054769</td>\n",
       "      <td>2</td>\n",
       "      <td>55</td>\n",
       "      <td>1.0</td>\n",
       "      <td>496</td>\n",
       "      <td>550.0</td>\n",
       "      <td>0.001818</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>653 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    species_name replicon_accession   k  total_k_mers  common_k_mers  \\\n",
       "0       Aardvark          NW_006921   2         11830          163.0   \n",
       "1       Aardvark          NW_006921   3         11702          102.0   \n",
       "2       Aardvark          NW_006921   5         10970           74.0   \n",
       "3       Aardvark          NW_006921   7         10266           62.0   \n",
       "4       Aardvark          NW_006921  11          9018           43.0   \n",
       "..           ...                ...  ..           ...            ...   \n",
       "648  Zebra Finch          NC_044234   2           159            4.0   \n",
       "649  Zebra Finch          NC_044234   3           158            3.0   \n",
       "650  Zebra Finch          NC_044234   5           156            1.0   \n",
       "651  Zebra Finch          NC_044236   2           144            1.0   \n",
       "652  Zebra Finch          NC_054769   2            55            1.0   \n",
       "\n",
       "     total_k_mers_origin  generated_k_mers  probability  \n",
       "0                    496           12163.0     0.013401  \n",
       "1                    508           12108.0     0.008424  \n",
       "2                    516           11412.0     0.006484  \n",
       "3                    520           10724.0     0.005781  \n",
       "4                    520            9495.0     0.004529  \n",
       "..                   ...               ...          ...  \n",
       "648                  496             651.0     0.006144  \n",
       "649                  508             663.0     0.004525  \n",
       "650                  516             671.0     0.001490  \n",
       "651                  496             639.0     0.001565  \n",
       "652                  496             550.0     0.001818  \n",
       "\n",
       "[653 rows x 8 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k_mers_counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For computing the metric we need:\n",
    "- Probability = coomon k-mers / generated k-mers between two scafffolds / species\n",
    "- Length of each scaffold\n",
    "- k of the k-mer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species_name</th>\n",
       "      <th>replicon_accession</th>\n",
       "      <th>k</th>\n",
       "      <th>total_k_mers</th>\n",
       "      <th>common_k_mers</th>\n",
       "      <th>total_k_mers_origin</th>\n",
       "      <th>generated_k_mers</th>\n",
       "      <th>probability</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aardvark</td>\n",
       "      <td>NW_006921</td>\n",
       "      <td>6469693230</td>\n",
       "      <td>-1943436426934344576</td>\n",
       "      <td>5.234950e+14</td>\n",
       "      <td>-3313660463278981120</td>\n",
       "      <td>3.835701e+39</td>\n",
       "      <td>1.364796e-25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Aardvark</td>\n",
       "      <td>NW_006922</td>\n",
       "      <td>210</td>\n",
       "      <td>4903958184840</td>\n",
       "      <td>9.918000e+03</td>\n",
       "      <td>67608053760</td>\n",
       "      <td>1.588917e+13</td>\n",
       "      <td>6.241989e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054371.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1766</td>\n",
       "      <td>4.000000e+00</td>\n",
       "      <td>496</td>\n",
       "      <td>2.258000e+03</td>\n",
       "      <td>1.771479e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054372.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1342</td>\n",
       "      <td>4.000000e+00</td>\n",
       "      <td>496</td>\n",
       "      <td>1.834000e+03</td>\n",
       "      <td>2.181025e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>African clawed frog</td>\n",
       "      <td>NC_054374.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1254</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>496</td>\n",
       "      <td>1.749000e+03</td>\n",
       "      <td>5.717553e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>335</th>\n",
       "      <td>Zebra Finch</td>\n",
       "      <td>NC_044232</td>\n",
       "      <td>2310</td>\n",
       "      <td>1732798217700</td>\n",
       "      <td>1.947456e+06</td>\n",
       "      <td>35156187955200</td>\n",
       "      <td>2.570383e+14</td>\n",
       "      <td>7.576522e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336</th>\n",
       "      <td>Zebra Finch</td>\n",
       "      <td>NC_044233</td>\n",
       "      <td>2</td>\n",
       "      <td>207</td>\n",
       "      <td>2.000000e+00</td>\n",
       "      <td>496</td>\n",
       "      <td>7.010000e+02</td>\n",
       "      <td>2.853067e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>337</th>\n",
       "      <td>Zebra Finch</td>\n",
       "      <td>NC_044234</td>\n",
       "      <td>30</td>\n",
       "      <td>3919032</td>\n",
       "      <td>1.200000e+01</td>\n",
       "      <td>130015488</td>\n",
       "      <td>2.896123e+08</td>\n",
       "      <td>4.143470e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338</th>\n",
       "      <td>Zebra Finch</td>\n",
       "      <td>NC_044236</td>\n",
       "      <td>2</td>\n",
       "      <td>144</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>496</td>\n",
       "      <td>6.390000e+02</td>\n",
       "      <td>1.564945e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339</th>\n",
       "      <td>Zebra Finch</td>\n",
       "      <td>NC_054769</td>\n",
       "      <td>2</td>\n",
       "      <td>55</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>496</td>\n",
       "      <td>5.500000e+02</td>\n",
       "      <td>1.818182e-03</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>340 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            species_name replicon_accession           k         total_k_mers  \\\n",
       "0               Aardvark          NW_006921  6469693230 -1943436426934344576   \n",
       "1               Aardvark          NW_006922         210        4903958184840   \n",
       "2    African clawed frog        NC_054371.1           2                 1766   \n",
       "3    African clawed frog        NC_054372.1           2                 1342   \n",
       "4    African clawed frog        NC_054374.1           2                 1254   \n",
       "..                   ...                ...         ...                  ...   \n",
       "335          Zebra Finch          NC_044232        2310        1732798217700   \n",
       "336          Zebra Finch          NC_044233           2                  207   \n",
       "337          Zebra Finch          NC_044234          30              3919032   \n",
       "338          Zebra Finch          NC_044236           2                  144   \n",
       "339          Zebra Finch          NC_054769           2                   55   \n",
       "\n",
       "     common_k_mers  total_k_mers_origin  generated_k_mers   probability  \n",
       "0     5.234950e+14 -3313660463278981120      3.835701e+39  1.364796e-25  \n",
       "1     9.918000e+03          67608053760      1.588917e+13  6.241989e-10  \n",
       "2     4.000000e+00                  496      2.258000e+03  1.771479e-03  \n",
       "3     4.000000e+00                  496      1.834000e+03  2.181025e-03  \n",
       "4     1.000000e+00                  496      1.749000e+03  5.717553e-04  \n",
       "..             ...                  ...               ...           ...  \n",
       "335   1.947456e+06       35156187955200      2.570383e+14  7.576522e-09  \n",
       "336   2.000000e+00                  496      7.010000e+02  2.853067e-03  \n",
       "337   1.200000e+01            130015488      2.896123e+08  4.143470e-08  \n",
       "338   1.000000e+00                  496      6.390000e+02  1.564945e-03  \n",
       "339   1.000000e+00                  496      5.500000e+02  1.818182e-03  \n",
       "\n",
       "[340 rows x 8 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metric"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''# Get the k-mers of the scaffold\n",
    "df_scaffold = df_list[df_list['replicon_accession'] == scaffold]\n",
    "\n",
    "# Get the k-mers of the rest of scaffolds\n",
    "#df_rest = df_list[df_list['replicon_accession'] != scaffold]\n",
    "\n",
    "# Get the k-mers shared with the rest of scaffolds\n",
    "df_shared = df_scaffold.merge(df_list, on = ['k', 'k_mer'])\n",
    "\n",
    "df_shared.rename(columns = {\n",
    "                    'species_name_y': 'species_name', \n",
    "                    'replicon_accession_y': 'replicon_accession'}, inplace = True)\n",
    "\n",
    "# Remove the k-mers shared within the same species\n",
    "df_shared = df_shared.loc[df_shared['species_name_x'] != df_shared['species_name'], ['k', 'species_name', 'replicon_accession', 'k_mer']]\n",
    "                    \n",
    "# Get the number of k-mers shared with the rest of scaffolds\n",
    "df_both = df_shared.groupby(['k', 'species_name', 'replicon_accession'], as_index=False).nunique().sort_values('k_mer', ascending = False)\n",
    "\n",
    "df_both = df_both[['k', 'species_name', 'replicon_accession', 'k_mer']].rename(columns = {'k_mer': 'common_k_mers'})'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''# Compute the number of k_mers per k per scaffold / species, and the number of k_mers shared with the rest of scaffolds\n",
    "# Then, do # scaffold k_mers + # match_scaffold k_mers - # shared k_mers to get the number of shared k_mers\n",
    "def generated_k_mers(scaffold, match_scaffold, df_k_mers):\n",
    "    # Create a list of tuples of k and k_mers per each scaffold\n",
    "    k_mers_scaffold = df_k_mers[df_k_mers['replicon_accession'] == scaffold][['k', 'k_mer']]\n",
    "    k_mers_match_scaffold = df_k_mers[df_k_mers['replicon_accession'] == match_scaffold][['k', 'k_mer']]\n",
    "\n",
    "    # Get the k-mers generated between the two scaffolds\n",
    "    generated_k_mers = pd.concat([k_mers_scaffold, k_mers_match_scaffold]).drop_duplicates()\n",
    "\n",
    "    # Get the number of k-mers generated between the two scaffolds per k\n",
    "    generated_k_mers = generated_k_mers.groupby(['k'], as_index=False).count()\n",
    "\n",
    "    generated_k_mers['replicon_accession'] = match_scaffold\n",
    "    generated_k_mers.rename(columns = {'k_mer': 'total_k_mers'}, inplace = True)\n",
    "\n",
    "    return generated_k_mers'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[13], line 8\u001b[0m\n\u001b[0;32m      5\u001b[0m df_total_k_mers \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mDataFrame()\n\u001b[0;32m      7\u001b[0m \u001b[39mfor\u001b[39;00m match_scaffold \u001b[39min\u001b[39;00m matching_scaffolds:\n\u001b[1;32m----> 8\u001b[0m     k_mers \u001b[39m=\u001b[39m df_list[(df_list[\u001b[39m'\u001b[39m\u001b[39mreplicon_accession\u001b[39m\u001b[39m'\u001b[39m] \u001b[39m==\u001b[39m scaffold) \u001b[39m|\u001b[39m (df_list[\u001b[39m'\u001b[39;49m\u001b[39mreplicon_accession\u001b[39;49m\u001b[39m'\u001b[39;49m] \u001b[39m==\u001b[39;49m match_scaffold)]\n\u001b[0;32m      9\u001b[0m     total_k_mers \u001b[39m=\u001b[39m k_mers\u001b[39m.\u001b[39mgroupby([\u001b[39m'\u001b[39m\u001b[39mk\u001b[39m\u001b[39m'\u001b[39m], as_index\u001b[39m=\u001b[39m\u001b[39mFalse\u001b[39;00m)\u001b[39m.\u001b[39mnunique()\u001b[39m.\u001b[39msort_values(\u001b[39m'\u001b[39m\u001b[39mk_mer\u001b[39m\u001b[39m'\u001b[39m, ascending \u001b[39m=\u001b[39m \u001b[39mFalse\u001b[39;00m)\n\u001b[0;32m     10\u001b[0m     total_k_mers[\u001b[39m'\u001b[39m\u001b[39mreplicon_accession\u001b[39m\u001b[39m'\u001b[39m] \u001b[39m=\u001b[39m match_scaffold\n",
      "File \u001b[1;32mc:\\Users\\bertr\\miniconda3\\envs\\thesis\\lib\\site-packages\\pandas\\core\\ops\\common.py:72\u001b[0m, in \u001b[0;36m_unpack_zerodim_and_defer.<locals>.new_method\u001b[1;34m(self, other)\u001b[0m\n\u001b[0;32m     68\u001b[0m             \u001b[39mreturn\u001b[39;00m \u001b[39mNotImplemented\u001b[39m\n\u001b[0;32m     70\u001b[0m other \u001b[39m=\u001b[39m item_from_zerodim(other)\n\u001b[1;32m---> 72\u001b[0m \u001b[39mreturn\u001b[39;00m method(\u001b[39mself\u001b[39;49m, other)\n",
      "File \u001b[1;32mc:\\Users\\bertr\\miniconda3\\envs\\thesis\\lib\\site-packages\\pandas\\core\\arraylike.py:42\u001b[0m, in \u001b[0;36mOpsMixin.__eq__\u001b[1;34m(self, other)\u001b[0m\n\u001b[0;32m     40\u001b[0m \u001b[39m@unpack_zerodim_and_defer\u001b[39m(\u001b[39m\"\u001b[39m\u001b[39m__eq__\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m     41\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m__eq__\u001b[39m(\u001b[39mself\u001b[39m, other):\n\u001b[1;32m---> 42\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_cmp_method(other, operator\u001b[39m.\u001b[39;49meq)\n",
      "File \u001b[1;32mc:\\Users\\bertr\\miniconda3\\envs\\thesis\\lib\\site-packages\\pandas\\core\\series.py:6243\u001b[0m, in \u001b[0;36mSeries._cmp_method\u001b[1;34m(self, other, op)\u001b[0m\n\u001b[0;32m   6240\u001b[0m rvalues \u001b[39m=\u001b[39m extract_array(other, extract_numpy\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m, extract_range\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m)\n\u001b[0;32m   6242\u001b[0m \u001b[39mwith\u001b[39;00m np\u001b[39m.\u001b[39merrstate(\u001b[39mall\u001b[39m\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mignore\u001b[39m\u001b[39m\"\u001b[39m):\n\u001b[1;32m-> 6243\u001b[0m     res_values \u001b[39m=\u001b[39m ops\u001b[39m.\u001b[39;49mcomparison_op(lvalues, rvalues, op)\n\u001b[0;32m   6245\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_construct_result(res_values, name\u001b[39m=\u001b[39mres_name)\n",
      "File \u001b[1;32mc:\\Users\\bertr\\miniconda3\\envs\\thesis\\lib\\site-packages\\pandas\\core\\ops\\array_ops.py:287\u001b[0m, in \u001b[0;36mcomparison_op\u001b[1;34m(left, right, op)\u001b[0m\n\u001b[0;32m    284\u001b[0m     \u001b[39mreturn\u001b[39;00m invalid_comparison(lvalues, rvalues, op)\n\u001b[0;32m    286\u001b[0m \u001b[39melif\u001b[39;00m is_object_dtype(lvalues\u001b[39m.\u001b[39mdtype) \u001b[39mor\u001b[39;00m \u001b[39misinstance\u001b[39m(rvalues, \u001b[39mstr\u001b[39m):\n\u001b[1;32m--> 287\u001b[0m     res_values \u001b[39m=\u001b[39m comp_method_OBJECT_ARRAY(op, lvalues, rvalues)\n\u001b[0;32m    289\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m    290\u001b[0m     res_values \u001b[39m=\u001b[39m _na_arithmetic_op(lvalues, rvalues, op, is_cmp\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m)\n",
      "File \u001b[1;32mc:\\Users\\bertr\\miniconda3\\envs\\thesis\\lib\\site-packages\\pandas\\core\\ops\\array_ops.py:76\u001b[0m, in \u001b[0;36mcomp_method_OBJECT_ARRAY\u001b[1;34m(op, x, y)\u001b[0m\n\u001b[0;32m     74\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m     75\u001b[0m     result \u001b[39m=\u001b[39m libops\u001b[39m.\u001b[39mscalar_compare(x\u001b[39m.\u001b[39mravel(), y, op)\n\u001b[1;32m---> 76\u001b[0m \u001b[39mreturn\u001b[39;00m result\u001b[39m.\u001b[39;49mreshape(x\u001b[39m.\u001b[39mshape)\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# Now, we want to compute the total number of k-mers that two scaffolds generate\n",
    "# One scaffold is given; the other one is the rest of scaffolds appearing in df_shared\n",
    "matching_scaffolds = df_both['replicon_accession'].unique()\n",
    "\n",
    "df_total_k_mers = pd.DataFrame()\n",
    "\n",
    "for match_scaffold in matching_scaffolds:\n",
    "    k_mers = df_list[(df_list['replicon_accession'] == scaffold) | (df_list['replicon_accession'] == match_scaffold)]\n",
    "    total_k_mers = k_mers.groupby(['k'], as_index=False).nunique().sort_values('k_mer', ascending = False)\n",
    "    total_k_mers['replicon_accession'] = match_scaffold\n",
    "    total_k_mers.rename(columns = {'k_mer': 'total_k_mers'}, inplace = True)\n",
    "\n",
    "    df_total_k_mers = pd.concat([df_total_k_mers, total_k_mers], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\bertr\\AppData\\Local\\Temp\\ipykernel_9244\\167985797.py:16: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  total_k_mers = total_k_mers[['k_mer']].rename(columns = {'k_mer': 'total_k_mers'}, inplace = True)\n",
      "C:\\Users\\bertr\\AppData\\Local\\Temp\\ipykernel_9244\\167985797.py:16: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  total_k_mers = total_k_mers[['k_mer']].rename(columns = {'k_mer': 'total_k_mers'}, inplace = True)\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "for match_scaffold in matching_scaffolds:\n",
    "    df_list['scaffold1'] = np.where(df_list['replicon_accession'] == scaffold, 1, 0)\n",
    "    df_list['scaffold2'] = np.where(df_list['replicon_accession'] == match_scaffold, 1, 0)\n",
    "\n",
    "    df_total = df_list.groupby(['k', 'k_mer'], as_index=False).max().sort_values('k_mer', ascending = False)\n",
    "    df_total['sum'] = df_total['scaffold1'] + df_total['scaffold2']\n",
    "\n",
    "    # Just take the k-mers that appear in both scaffolds\n",
    "    df_total = df_total[df_total['sum'] >= 1]\n",
    "\n",
    "    # unique is not needed, since are already grouped by k and k_mer\n",
    "    total_k_mers = df_total.groupby(['k'], as_index=False).count().sort_values('k_mer', ascending = False)\n",
    "    total_k_mers['replicon_accession'] = match_scaffold\n",
    "    total_k_mers = total_k_mers[['k_mer']].rename(columns = {'k_mer': 'total_k_mers'}, inplace = True)\n",
    "\n",
    "    df_total_k_mers = pd.concat([df_total_k_mers, total_k_mers], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "num_k_mers = {}\n",
    "\n",
    "for match_scaffold in matching_scaffolds:\n",
    "\n",
    "    # Create a list of tuples of k and k_mers per each scaffold\n",
    "    k_mers_scaffold = [tuple(row) for row in df_list[df_list['replicon_accession'] == scaffold][['k', 'k_mer']].values.tolist()]\n",
    "    k_mers_match_scaffold = [tuple(row) for row in df_list[df_list['replicon_accession'] == match_scaffold][['k', 'k_mer']].values.tolist()]\n",
    "\n",
    "    # Get the k-mers generates between the two scaffolds\t\n",
    "    generated_k_mers = set(k_mers_scaffold).union(k_mers_match_scaffold)\n",
    "\n",
    "    # Get the number of k-mers generated between the two scaffolds per k\n",
    "    generated_k_mers = {k: len(v) for k, v in generated_k_mers}\n",
    "\n",
    "    # Add the scaffold name to the dictionary\n",
    "    num_k_mers[match_scaffold] = generated_k_mers\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for match_scaffold in matching_scaffolds:\n",
    "\n",
    "    # Create a list of tuples of k and k_mers per each scaffold\n",
    "    k_mers_scaffold = df_list[df_list['replicon_accession'] == scaffold][['k', 'k_mer']]\n",
    "    k_mers_match_scaffold = df_list[df_list['replicon_accession'] == match_scaffold][['k', 'k_mer']]\n",
    "\n",
    "    # Get the k-mers generated between the two scaffolds\n",
    "    generated_k_mers = pd.concat([k_mers_scaffold, k_mers_match_scaffold]).drop_duplicates()\n",
    "\n",
    "    # Get the number of k-mers generated between the two scaffolds per k\n",
    "    generated_k_mers = generated_k_mers.groupby(['k'], as_index=False).count()\n",
    "\n",
    "    generated_k_mers['replicon_accession'] = match_scaffold\n",
    "    generated_k_mers.rename(columns = {'k_mer': 'total_k_mers'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>k</th>\n",
       "      <th>k_mer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "      <td>845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>851</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7</td>\n",
       "      <td>853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11</td>\n",
       "      <td>849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>13</td>\n",
       "      <td>847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>17</td>\n",
       "      <td>843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>19</td>\n",
       "      <td>841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>23</td>\n",
       "      <td>833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>29</td>\n",
       "      <td>821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>31</td>\n",
       "      <td>817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>37</td>\n",
       "      <td>805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>41</td>\n",
       "      <td>797</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>43</td>\n",
       "      <td>793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>47</td>\n",
       "      <td>785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>53</td>\n",
       "      <td>773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>59</td>\n",
       "      <td>761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>61</td>\n",
       "      <td>757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>67</td>\n",
       "      <td>745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>71</td>\n",
       "      <td>737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>73</td>\n",
       "      <td>733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>79</td>\n",
       "      <td>721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>83</td>\n",
       "      <td>713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>89</td>\n",
       "      <td>701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>97</td>\n",
       "      <td>685</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     k  k_mer\n",
       "0    2    831\n",
       "1    3    845\n",
       "2    5    851\n",
       "3    7    853\n",
       "4   11    849\n",
       "5   13    847\n",
       "6   17    843\n",
       "7   19    841\n",
       "8   23    833\n",
       "9   29    821\n",
       "10  31    817\n",
       "11  37    805\n",
       "12  41    797\n",
       "13  43    793\n",
       "14  47    785\n",
       "15  53    773\n",
       "16  59    761\n",
       "17  61    757\n",
       "18  67    745\n",
       "19  71    737\n",
       "20  73    733\n",
       "21  79    721\n",
       "22  83    713\n",
       "23  89    701\n",
       "24  97    685"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "generated_k_mers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for match_scaffold in matching_scaffolds:\n",
    "\n",
    "    # Create a list of tuples of k and k_mers per each scaffold\n",
    "    k_mers_scaffold = df_list[df_list['replicon_accession'] == scaffold][['k', 'k_mer']]\n",
    "    k_mers_match_scaffold = df_list[df_list['replicon_accession'] == match_scaffold][['k', 'k_mer']]\n",
    "\n",
    "    # Get the k-mers generated between the two scaffolds\n",
    "    generated_k_mers = k_mers_scaffold.join(k_mers_match_scaffold, lsuffix='_scaffold', rsuffix='_match_scaffold', how='outer')\n",
    "\n",
    "    # Get the number of k-mers generated between the two scaffolds per k\n",
    "    generated_k_mers = generated_k_mers.groupby(['k'], as_index=False).count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for match_scaffold in matching_scaffolds:\n",
    "    # Create a list of tuples of k and k_mers per each scaffold\n",
    "    k_mers_scaffold = df_list[df_list['replicon_accession'] == scaffold][['k', 'k_mer']]\n",
    "    k_mers_match_scaffold = df_list[df_list['replicon_accession'] == match_scaffold][['k', 'k_mer']]\n",
    "\n",
    "    # Get the k-mers generated between the two scaffolds\n",
    "    generated_k_mers = k_mers_scaffold.join(k_mers_match_scaffold, lsuffix='_scaffold', rsuffix='_match_scaffold', how='outer')\n",
    "\n",
    "    # Get the number of k-mers generated between the two scaffolds per k\n",
    "    generated_k_mers = generated_k_mers.groupby(['k_scaffold'], as_index=False).count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "match_scaffold = 'NZ_CP010000.1'\n",
    "\n",
    "# Create a list of tuples of k and k_mers per each scaffold\n",
    "k_mers_scaffold = df_list[df_list['replicon_accession'] == scaffold][['k', 'k_mer']]\n",
    "k_mers_match_scaffold = df_list[df_list['replicon_accession'] == match_scaffold][['k', 'k_mer']]\n",
    "\n",
    "# Get the k-mers generated between the two scaffolds\n",
    "generated_k_mers = k_mers_scaffold.join(k_mers_match_scaffold, lsuffix='_scaffold', rsuffix='_match_scaffold', how='outer')\n",
    "\n",
    "# Get the number of k-mers generated between the two scaffolds per k\n",
    "generated_k_mers = generated_k_mers.groupby(['k_scaffold'], as_index=False).count()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "thesis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
