{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import metric as me\n",
    "import swco\n",
    "import k_mers_metric as kmm"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the k_mers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "101 :  101_mers.csv\n",
      "103 :  103_mers.csv\n",
      "107 :  107_mers.csv\n",
      "109 :  109_mers.csv\n",
      "113 :  113_mers.csv\n",
      "11 :  11_mers.csv\n",
      "127 :  127_mers.csv\n",
      "131 :  131_mers.csv\n",
      "137 :  137_mers.csv\n",
      "139 :  139_mers.csv\n",
      "13 :  13_mers.csv\n",
      "149 :  149_mers.csv\n",
      "151 :  151_mers.csv\n",
      "157 :  157_mers.csv\n",
      "163 :  163_mers.csv\n",
      "167 :  167_mers.csv\n",
      "173 :  173_mers.csv\n",
      "179 :  179_mers.csv\n",
      "17 :  17_mers.csv\n",
      "181 :  181_mers.csv\n",
      "191 :  191_mers.csv\n",
      "193 :  193_mers.csv\n",
      "197 :  197_mers.csv\n",
      "199 :  199_mers.csv\n",
      "19 :  19_mers.csv\n",
      "211 :  211_mers.csv\n",
      "223 :  223_mers.csv\n",
      "227 :  227_mers.csv\n",
      "229 :  229_mers.csv\n",
      "233 :  233_mers.csv\n",
      "239 :  239_mers.csv\n",
      "23 :  23_mers.csv\n",
      "241 :  241_mers.csv\n",
      "29 :  29_mers.csv\n",
      "2 :  2_mers.csv\n",
      "31 :  31_mers.csv\n",
      "37 :  37_mers.csv\n",
      "3 :  3_mers.csv\n",
      "41 :  41_mers.csv\n",
      "43 :  43_mers.csv\n",
      "47 :  47_mers.csv\n",
      "53 :  53_mers.csv\n",
      "59 :  59_mers.csv\n",
      "5 :  5_mers.csv\n",
      "61 :  61_mers.csv\n",
      "67 :  67_mers.csv\n",
      "71 :  71_mers.csv\n",
      "73 :  73_mers.csv\n",
      "79 :  79_mers.csv\n",
      "7 :  7_mers.csv\n",
      "83 :  83_mers.csv\n",
      "89 :  89_mers.csv\n",
      "97 :  97_mers.csv\n"
     ]
    }
   ],
   "source": [
    "# Just the ones with prime k values\n",
    "#k_mers = me.read_k_mers_prime('../Data/Intermediate/k_mers/Scaffold')\n",
    "\n",
    "# Version to read all of them: me.read_k_mers(path), not only for k prime numbers\n",
    "\n",
    "# aprox. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_k_mers = pd.read_feather('../Data/Intermediate/k_mers/processed/df_k_mers_relaxed.feather')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"# Let's read the new k-mers from GFF files\\nimport read_k_mers as rkm\\n\\n# Path to the folder with the processed files\\npath = '../Data/Intermediate/k_mers/processed/'\\ndata_list = rkm.process_folder(path)\\n\\n# Convert the list of dictionaries into a single DataFrame, including the species and k columns\\ndf_list = pd.concat([d['data'].assign(species_gcf_id=d['species_gcf_id'], k=d['k']) for d in data_list], ignore_index=True)\\n#df_list = pd.concat([d['data'] for d in data_list], ignore_index=True)\\ndf_list = df_list[['species_gcf_id', 'k', 'k_mer', 'replicon_accession']].rename(columns={'replicon_accession': 'scaffold'})\\n\\n# Create a column in the dataframe with the species name\\nfile_species = rkm.map_species_to_file()\\ndf_list['species'] = df_list['species_gcf_id'].map(file_species)\""
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''# Let's read the new k-mers from GFF files\n",
    "import read_k_mers as rkm\n",
    "\n",
    "# Path to the folder with the processed files\n",
    "path = '../Data/Intermediate/k_mers/processed/'\n",
    "data_list = rkm.process_folder(path)\n",
    "\n",
    "# Convert the list of dictionaries into a single DataFrame, including the species and k columns\n",
    "df_list = pd.concat([d['data'].assign(species_gcf_id=d['species_gcf_id'], k=d['k']) for d in data_list], ignore_index=True)\n",
    "#df_list = pd.concat([d['data'] for d in data_list], ignore_index=True)\n",
    "df_list = df_list[['species_gcf_id', 'k', 'k_mer', 'replicon_accession']].rename(columns={'replicon_accession': 'scaffold'})\n",
    "\n",
    "# Create a column in the dataframe with the species name\n",
    "file_species = rkm.map_species_to_file()\n",
    "df_list['species'] = df_list['species_gcf_id'].map(file_species)'''"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the species and their lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Specie</th>\n",
       "      <th>Number of genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aadvark</td>\n",
       "      <td>14841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Alligator M</td>\n",
       "      <td>13596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Alligator S</td>\n",
       "      <td>13140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Anolis</td>\n",
       "      <td>12650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Chelonia</td>\n",
       "      <td>12293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Chrysemys</td>\n",
       "      <td>11218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Croco</td>\n",
       "      <td>12138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Danio</td>\n",
       "      <td>25221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Devil</td>\n",
       "      <td>14059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Dog</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Fugu</td>\n",
       "      <td>1733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Gecko</td>\n",
       "      <td>12708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Gorilla</td>\n",
       "      <td>15835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Human</td>\n",
       "      <td>18874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>KCobra</td>\n",
       "      <td>13474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Koala</td>\n",
       "      <td>14792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Monodelphis</td>\n",
       "      <td>14321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Mouse</td>\n",
       "      <td>18860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Orca</td>\n",
       "      <td>14997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Ostrich</td>\n",
       "      <td>11490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Papio</td>\n",
       "      <td>15930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Pelodiscus</td>\n",
       "      <td>12469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Pigeon</td>\n",
       "      <td>542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Platypus</td>\n",
       "      <td>10547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Pogona</td>\n",
       "      <td>13658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Python</td>\n",
       "      <td>12356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Thamnophis</td>\n",
       "      <td>10571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Xenopus</td>\n",
       "      <td>21457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Xlaevis</td>\n",
       "      <td>10976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Xtropicalis</td>\n",
       "      <td>14416</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Specie  Number of genes\n",
       "0       Aadvark            14841\n",
       "1   Alligator M            13596\n",
       "2   Alligator S            13140\n",
       "3        Anolis            12650\n",
       "4      Chelonia            12293\n",
       "5     Chrysemys            11218\n",
       "6         Croco            12138\n",
       "7         Danio            25221\n",
       "8         Devil            14059\n",
       "9           Dog               60\n",
       "10         Fugu             1733\n",
       "11        Gecko            12708\n",
       "12      Gorilla            15835\n",
       "13        Human            18874\n",
       "14       KCobra            13474\n",
       "15        Koala            14792\n",
       "16  Monodelphis            14321\n",
       "17        Mouse            18860\n",
       "18         Orca            14997\n",
       "19      Ostrich            11490\n",
       "20        Papio            15930\n",
       "21   Pelodiscus            12469\n",
       "22       Pigeon              542\n",
       "23     Platypus            10547\n",
       "24       Pogona            13658\n",
       "25       Python            12356\n",
       "26   Thamnophis            10571\n",
       "27      Xenopus            21457\n",
       "28      Xlaevis            10976\n",
       "29  Xtropicalis            14416"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_species = pd.read_csv('../Data/Intermediate/species_length.csv')        #sorted(df_species['Specie'].unique())\n",
    "df_species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\nWhen I have the new GFF files, compute the length of the scaffolds and the species\\nThen, we can create the measure matrix with this code\\n\\nimport k_mers_metric as kmm\\n\\n# Given a specific scaffold, know how many k-mers shared with the rest of scaffolds\\nlevel = 'species'\\n\\nk_mers_counter = pd.read_feather('../Data/Intermediate/k_mers/count_k_mers_' + level + '.feather')\\n\\ntry:\\n    df_length = pd.read_csv('../Data/Intermediate/' + level + '_length.csv')\\n    if level == 'scaffold':\\n        df_length = (df_length[['Specie', 'Replicon Accession', 'Number of genes']]\\n                        .rename(columns = {'Specie': 'species', 'Replicon Accession': 'scaffold', 'Number of genes': 'length'}))\\n    elif level == 'species':\\n        df_length = (df_length[['Specie', 'Number of genes']]\\n                        .rename(columns = {'Specie': 'species', 'Number of genes': 'length'}))\\nexcept:\\n    print('Please, check the columns of the file ' + level + '_length.csv')\\n\\nmeasure = pd.DataFrame()\\n\\nfor species in df_species['Specie']:\\n    print(species)\\n    species = kmm.k_mers_to_metric(species, level, k_mers_counter, df_length, df_k_mers)\\n    species['species1'] = origin\\n    species.rename(columns={'species2': 'species'}, inplace=True)\\n    \\n    measure = pd.concat([measure, species])\\n\""
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "When I have the new GFF files, compute the length of the scaffolds and the species\n",
    "Then, we can create the measure matrix with this code\n",
    "\n",
    "import k_mers_metric as kmm\n",
    "\n",
    "# Given a specific scaffold, know how many k-mers shared with the rest of scaffolds\n",
    "level = 'species'\n",
    "\n",
    "k_mers_counter = pd.read_feather('../Data/Intermediate/k_mers/count_k_mers_' + level + '.feather')\n",
    "\n",
    "try:\n",
    "    df_length = pd.read_csv('../Data/Intermediate/' + level + '_length.csv')\n",
    "    if level == 'scaffold':\n",
    "        df_length = (df_length[['Specie', 'Replicon Accession', 'Number of genes']]\n",
    "                        .rename(columns = {'Specie': 'species', 'Replicon Accession': 'scaffold', 'Number of genes': 'length'}))\n",
    "    elif level == 'species':\n",
    "        df_length = (df_length[['Specie', 'Number of genes']]\n",
    "                        .rename(columns = {'Specie': 'species', 'Number of genes': 'length'}))\n",
    "except:\n",
    "    print('Please, check the columns of the file ' + level + '_length.csv')\n",
    "\n",
    "measure = pd.DataFrame()\n",
    "\n",
    "for species in df_species['Specie']:\n",
    "    print(species)\n",
    "    species = kmm.k_mers_to_metric(species, level, k_mers_counter, df_length, df_k_mers)\n",
    "    species['species1'] = origin\n",
    "    species.rename(columns={'species2': 'species'}, inplace=True)\n",
    "    \n",
    "    measure = pd.concat([measure, species])\n",
    "'''\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Specie</th>\n",
       "      <th>Number of genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aadvark</td>\n",
       "      <td>14841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Alligator M</td>\n",
       "      <td>13596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Alligator S</td>\n",
       "      <td>13140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Anolis</td>\n",
       "      <td>12650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Chelonia</td>\n",
       "      <td>12293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Chrysemys</td>\n",
       "      <td>11218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Croco</td>\n",
       "      <td>12138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Danio</td>\n",
       "      <td>25221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Devil</td>\n",
       "      <td>14059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Dog</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Fugu</td>\n",
       "      <td>1733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Gecko</td>\n",
       "      <td>12708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Gorilla</td>\n",
       "      <td>15835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Human</td>\n",
       "      <td>18874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>KCobra</td>\n",
       "      <td>13474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Koala</td>\n",
       "      <td>14792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Monodelphis</td>\n",
       "      <td>14321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Mouse</td>\n",
       "      <td>18860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Orca</td>\n",
       "      <td>14997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Ostrich</td>\n",
       "      <td>11490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Papio</td>\n",
       "      <td>15930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Pelodiscus</td>\n",
       "      <td>12469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Pigeon</td>\n",
       "      <td>542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Platypus</td>\n",
       "      <td>10547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Pogona</td>\n",
       "      <td>13658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Python</td>\n",
       "      <td>12356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Thamnophis</td>\n",
       "      <td>10571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Xenopus</td>\n",
       "      <td>21457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Xlaevis</td>\n",
       "      <td>10976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Xtropicalis</td>\n",
       "      <td>14416</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Specie  Number of genes\n",
       "0       Aadvark            14841\n",
       "1   Alligator M            13596\n",
       "2   Alligator S            13140\n",
       "3        Anolis            12650\n",
       "4      Chelonia            12293\n",
       "5     Chrysemys            11218\n",
       "6         Croco            12138\n",
       "7         Danio            25221\n",
       "8         Devil            14059\n",
       "9           Dog               60\n",
       "10         Fugu             1733\n",
       "11        Gecko            12708\n",
       "12      Gorilla            15835\n",
       "13        Human            18874\n",
       "14       KCobra            13474\n",
       "15        Koala            14792\n",
       "16  Monodelphis            14321\n",
       "17        Mouse            18860\n",
       "18         Orca            14997\n",
       "19      Ostrich            11490\n",
       "20        Papio            15930\n",
       "21   Pelodiscus            12469\n",
       "22       Pigeon              542\n",
       "23     Platypus            10547\n",
       "24       Pogona            13658\n",
       "25       Python            12356\n",
       "26   Thamnophis            10571\n",
       "27      Xenopus            21457\n",
       "28      Xlaevis            10976\n",
       "29  Xtropicalis            14416"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "0",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[1;32mc:\\Users\\bertr\\miniconda3\\envs\\thesis\\lib\\site-packages\\pandas\\core\\indexes\\base.py:3802\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[1;34m(self, key, method, tolerance)\u001b[0m\n\u001b[0;32m   3801\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m-> 3802\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_engine\u001b[39m.\u001b[39;49mget_loc(casted_key)\n\u001b[0;32m   3803\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mKeyError\u001b[39;00m \u001b[39mas\u001b[39;00m err:\n",
      "File \u001b[1;32mc:\\Users\\bertr\\miniconda3\\envs\\thesis\\lib\\site-packages\\pandas\\_libs\\index.pyx:138\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mc:\\Users\\bertr\\miniconda3\\envs\\thesis\\lib\\site-packages\\pandas\\_libs\\index.pyx:165\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mpandas\\_libs\\hashtable_class_helper.pxi:5745\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mpandas\\_libs\\hashtable_class_helper.pxi:5753\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;31mKeyError\u001b[0m: 0",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[7], line 4\u001b[0m\n\u001b[0;32m      1\u001b[0m metric \u001b[39m=\u001b[39m []\n\u001b[0;32m      3\u001b[0m \u001b[39mfor\u001b[39;00m index_sp1 \u001b[39min\u001b[39;00m \u001b[39mrange\u001b[39m(\u001b[39mlen\u001b[39m(df_species)\u001b[39m-\u001b[39m\u001b[39m1\u001b[39m):\n\u001b[1;32m----> 4\u001b[0m     sp1, len_sp1 \u001b[39m=\u001b[39m df_species[index_sp1]\n\u001b[0;32m      6\u001b[0m     \u001b[39mfor\u001b[39;00m index_sp2 \u001b[39min\u001b[39;00m \u001b[39mrange\u001b[39m(index_sp1\u001b[39m+\u001b[39m\u001b[39m1\u001b[39m, \u001b[39mlen\u001b[39m(df_species)):\n\u001b[0;32m      7\u001b[0m         sp2, len_sp2 \u001b[39m=\u001b[39m df_species[index_sp2]\n",
      "File \u001b[1;32mc:\\Users\\bertr\\miniconda3\\envs\\thesis\\lib\\site-packages\\pandas\\core\\frame.py:3807\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m   3805\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mcolumns\u001b[39m.\u001b[39mnlevels \u001b[39m>\u001b[39m \u001b[39m1\u001b[39m:\n\u001b[0;32m   3806\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_getitem_multilevel(key)\n\u001b[1;32m-> 3807\u001b[0m indexer \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mcolumns\u001b[39m.\u001b[39;49mget_loc(key)\n\u001b[0;32m   3808\u001b[0m \u001b[39mif\u001b[39;00m is_integer(indexer):\n\u001b[0;32m   3809\u001b[0m     indexer \u001b[39m=\u001b[39m [indexer]\n",
      "File \u001b[1;32mc:\\Users\\bertr\\miniconda3\\envs\\thesis\\lib\\site-packages\\pandas\\core\\indexes\\base.py:3804\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[1;34m(self, key, method, tolerance)\u001b[0m\n\u001b[0;32m   3802\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_engine\u001b[39m.\u001b[39mget_loc(casted_key)\n\u001b[0;32m   3803\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mKeyError\u001b[39;00m \u001b[39mas\u001b[39;00m err:\n\u001b[1;32m-> 3804\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mKeyError\u001b[39;00m(key) \u001b[39mfrom\u001b[39;00m \u001b[39merr\u001b[39;00m\n\u001b[0;32m   3805\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mTypeError\u001b[39;00m:\n\u001b[0;32m   3806\u001b[0m     \u001b[39m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[0;32m   3807\u001b[0m     \u001b[39m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[0;32m   3808\u001b[0m     \u001b[39m#  the TypeError.\u001b[39;00m\n\u001b[0;32m   3809\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[1;31mKeyError\u001b[0m: 0"
     ]
    }
   ],
   "source": [
    "metric = []\n",
    "\n",
    "for index_sp1 in range(len(df_species)-1):\n",
    "    sp1, len_sp1 = df_species[index_sp1]\n",
    "\n",
    "    for index_sp2 in range(index_sp1+1, len(df_species)):\n",
    "        sp2, len_sp2 = df_species[index_sp2]\n",
    "\n",
    "        current = { \n",
    "                    'Specie1': sp1,\n",
    "                    'Specie2': sp2,\n",
    "                    'Ratio': me.metric_comp(sp1, sp2, k_mers, len_sp1, len_sp2)\n",
    "                }\n",
    "        print(current)\n",
    "        \n",
    "        metric.append(current)\n",
    "metric = pd.DataFrame(metric)\n",
    "\n",
    "#148 min (with iterruption)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "module 'metric' has no attribute 'metric_to_matrix'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[32], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m me\u001b[39m.\u001b[39;49mmetric_to_matrix(metric)\u001b[39m.\u001b[39mto_csv(\u001b[39m'\u001b[39m\u001b[39m../Data/Results/Metric/metric_173_prime_m_n.csv\u001b[39m\u001b[39m'\u001b[39m)\n",
      "\u001b[1;31mAttributeError\u001b[0m: module 'metric' has no attribute 'metric_to_matrix'"
     ]
    }
   ],
   "source": [
    "me.metric_to_matrix(metric).to_csv('../Data/Intermediate/metric_matrix_relaxed_table.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Transform it into a distance\n",
    "#me.metric_to_matrix(metric[['Specie1', 'Specie2', 'Distance_1']]).to_csv('../Data/Results/Metric/Distance/distance_1.csv', index=False)\n",
    "#me.metric_to_matrix(metric[['Specie1', 'Specie2', 'Distance_1/2']]).to_csv('../Data/Results/Metric/Distance/distance_12.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "thesis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
