{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "99a4b24e-557f-45b6-b542-e7581de69716",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Smith-Waterman local alignmentimplementation\n",
    "Adapted application of Smith-Watermann algorithm. We apply this algorithm to genes, instead of aminoacid letters.\n",
    "\n",
    "<i>Source: https://github.com/slavianap/Smith-Waterman-Algorithm#readme"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "644914f9-2043-4b83-b576-54bc24f6cd42",
   "metadata": {},
   "outputs": [],
   "source": [
    "# -*- coding: utf-8 -*-\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "LIFE97011 - Computing\n",
    "Python Programming - Assessed Exercise No. 3\n",
    "Task: Smith Waterman local alignment\n",
    "@Author: Slaviana Pavlovich\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "# Importing Python packages\n",
    "from enum import IntEnum\n",
    "import numpy as np\n",
    "\n",
    "# Assigning the constants for the scores\n",
    "class Score(IntEnum):\n",
    "    MATCH = 1\n",
    "    MISMATCH = -1\n",
    "    GAP = -1\n",
    "\n",
    "# Assigning the constant values for the traceback\n",
    "class Trace(IntEnum):\n",
    "    STOP = 0\n",
    "    LEFT = 1 \n",
    "    UP = 2\n",
    "    DIAGONAL = 3\n",
    "\n",
    "# Reading the fasta file and keeping the formatted sequence's name and sequence\n",
    "'''\n",
    "def fasta_reader(sequence_file):\n",
    "    lines = open(sequence_file).readlines()\n",
    "    sequence_name_row = lines[0][1:]\n",
    "    sequence = lines[1]\n",
    "    return sequence_name_row.replace(\" \", \"\").strip(), sequence.strip()\n",
    "'''\n",
    "\n",
    "# Implementing the Smith Waterman local alignment\n",
    "def smith_waterman(seq1, seq2):\n",
    "    # Generating the empty matrices for storing scores and tracing\n",
    "    row = len(seq1) + 1\n",
    "    col = len(seq2) + 1\n",
    "    matrix = np.zeros(shape=(row, col), dtype=np.int)  \n",
    "    tracing_matrix = np.zeros(shape=(row, col), dtype=np.int)  \n",
    "    \n",
    "    # Initialising the variables to find the highest scoring cell\n",
    "    max_score = -1\n",
    "    max_index = (-1, -1)\n",
    "    \n",
    "    # Calculating the scores for all cells in the matrix\n",
    "    for i in range(1, row):\n",
    "        for j in range(1, col):\n",
    "            # Calculating the diagonal score (match score)\n",
    "            match_value = Score.MATCH if seq1[i - 1] == seq2[j - 1] else Score.MISMATCH\n",
    "            diagonal_score = matrix[i - 1, j - 1] + match_value\n",
    "            \n",
    "            # Calculating the vertical gap score\n",
    "            vertical_score = matrix[i - 1, j] + Score.GAP\n",
    "            \n",
    "            # Calculating the horizontal gap score\n",
    "            horizontal_score = matrix[i, j - 1] + Score.GAP\n",
    "            \n",
    "            # Taking the highest score \n",
    "            matrix[i, j] = max(0, diagonal_score, vertical_score, horizontal_score)\n",
    "            \n",
    "            # Tracking where the cell's value is coming from    \n",
    "            if matrix[i, j] == 0: \n",
    "                tracing_matrix[i, j] = Trace.STOP\n",
    "                \n",
    "            elif matrix[i, j] == horizontal_score: \n",
    "                tracing_matrix[i, j] = Trace.LEFT\n",
    "                \n",
    "            elif matrix[i, j] == vertical_score: \n",
    "                tracing_matrix[i, j] = Trace.UP\n",
    "                \n",
    "            elif matrix[i, j] == diagonal_score: \n",
    "                tracing_matrix[i, j] = Trace.DIAGONAL \n",
    "                \n",
    "            # Tracking the cell with the maximum score\n",
    "            if matrix[i, j] >= max_score:\n",
    "                max_index = (i,j)\n",
    "                max_score = matrix[i, j]\n",
    "    \n",
    "    # Initialising the variables for tracing\n",
    "    aligned_seq1 = [ ] #\"\"\n",
    "    aligned_seq2 = [ ] #\"\"   \n",
    "    current_aligned_seq1 = [ ] #\"\"   \n",
    "    current_aligned_seq2 = [ ] #\"\"  \n",
    "    (max_i, max_j) = max_index\n",
    "    \n",
    "    # Tracing and computing the pathway with the local alignment\n",
    "    while tracing_matrix[max_i, max_j] != Trace.STOP:\n",
    "        if tracing_matrix[max_i, max_j] == Trace.DIAGONAL:\n",
    "            current_aligned_seq1.append(seq1[max_i - 1])\n",
    "            current_aligned_seq2.append(seq2[max_j - 1])\n",
    "            max_i = max_i - 1\n",
    "            max_j = max_j - 1\n",
    "            \n",
    "        elif tracing_matrix[max_i, max_j] == Trace.UP:\n",
    "            current_aligned_seq1.append(seq1[max_i - 1])\n",
    "            current_aligned_seq2.append('-, ')\n",
    "            max_i = max_i - 1    \n",
    "            \n",
    "        elif tracing_matrix[max_i, max_j] == Trace.LEFT:\n",
    "            current_aligned_seq1.append('-, ')\n",
    "            current_aligned_seq2.append(seq2[max_j - 1])\n",
    "            max_j = max_j - 1\n",
    "            \n",
    "        #aligned_seq1 = aligned_seq1 + current_aligned_seq1\n",
    "        #aligned_seq2 = aligned_seq2 + current_aligned_seq2\n",
    "    \n",
    "    # Reversing the order of the sequences\n",
    "    current_aligned_seq1 = current_aligned_seq1[::-1]\n",
    "    current_aligned_seq2 = current_aligned_seq2[::-1]\n",
    "    \n",
    "    return current_aligned_seq1, current_aligned_seq2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d8cd4de",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\bertr\\AppData\\Local\\Temp\\ipykernel_4340\\3501219787.py:43: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.\n",
      "Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations\n",
      "  matrix = np.zeros(shape=(row, col), dtype=np.int)\n",
      "C:\\Users\\bertr\\AppData\\Local\\Temp\\ipykernel_4340\\3501219787.py:44: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.\n",
      "Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations\n",
      "  tracing_matrix = np.zeros(shape=(row, col), dtype=np.int)\n"
     ]
    }
   ],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.6",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "vscode": {
   "interpreter": {
    "hash": "0e92abe02844fb54062d758795a97b84494055d25efe9d020733c23e9f9966e7"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
