{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## K-mers\n",
    "\n",
    "A k-mer is a part of a string of longitude k. This substrings are created consequently. So, all the k-mers of a string are all the consecutive substrings contained in a string.\n",
    "The aim of this script is to generate a dictionary with all the possible k-mers we found among all the species. The dictionary have the possible k-mers as keys and the values will be the origin and the Scaffold where this k-mers have been found in the original string."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# To load all of them at once and afterwards accessing one by one, this might be an option.\n",
    "#https://towardsdatascience.com/a-simple-trick-to-load-multiple-excel-worksheets-in-pandas-3fae4124345b\n",
    "# Define filepath\n",
    "filepath = '../Data/Raw/Tables_Filtered_IK_format.xlsx'\n",
    "\n",
    "# Load Excel file using Pandas with `sheet_name=None`\n",
    "df_dict = pd.read_excel(filepath, sheet_name=None)\n",
    "\n",
    "# Preview\n",
    "#df_dict\n",
    "\n",
    "# Get a specific one\n",
    "#human = df_dict.get('Human')\n",
    "\n",
    "# aprox 3 min 40 secs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def preprocessing(specie):\n",
    "    \n",
    "    specie['Gene_non_or'] = specie['Locus'].str.split('(\\d+)').str[0]\n",
    "    specie['Gene'] = specie['Gene_non_or'] + specie['Strand']\n",
    "    specie['Origin_Scaffold'] = specie['Origin'] + '_' + specie['Replicon Accession']\n",
    "    #specie.reset_index(inplace= True)\n",
    "\n",
    "    specie = specie[(specie['Gene_non_or'].str.contains('LOC') == False) & (specie['Gene_non_or'].str.isspace() == False) & (specie['Gene_non_or'] != '-') ]  \n",
    "\n",
    "    return specie"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data cleaning for each specie\n",
    "df_species = pd.DataFrame()\n",
    "\n",
    "species = df_dict.keys()\n",
    "\n",
    "for s in species:\n",
    "    aux = df_dict.get(s)\n",
    "    aux['Origin'] = s\n",
    "    df_species = pd.concat([df_species, aux]) #= pd.concat(df_species, aux)\n",
    "\n",
    "#species = preprocessing(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_species = preprocessing(df_species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#Replicon Name</th>\n",
       "      <th>Replicon Accession</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Locus</th>\n",
       "      <th>Protein product</th>\n",
       "      <th>Length</th>\n",
       "      <th>Protein name</th>\n",
       "      <th>Origin</th>\n",
       "      <th>Unnamed: 10</th>\n",
       "      <th>Locus tag</th>\n",
       "      <th>Gene_non_or</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Origin_Scaffold</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Un</td>\n",
       "      <td>NW_001878251.4</td>\n",
       "      <td>61071</td>\n",
       "      <td>65540</td>\n",
       "      <td>+</td>\n",
       "      <td>563697</td>\n",
       "      <td>znf995</td>\n",
       "      <td>XP_017209820.1</td>\n",
       "      <td>358</td>\n",
       "      <td>gastrula zinc finger protein XlCGF8.2DB</td>\n",
       "      <td>Danio</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>znf</td>\n",
       "      <td>znf+</td>\n",
       "      <td>Danio_NW_001878251.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Un</td>\n",
       "      <td>NW_001884452.4</td>\n",
       "      <td>167019</td>\n",
       "      <td>169622</td>\n",
       "      <td>+</td>\n",
       "      <td>108183944</td>\n",
       "      <td>si:ch211-198g14.4</td>\n",
       "      <td>XP_021326470.1</td>\n",
       "      <td>266</td>\n",
       "      <td>actin-binding protein-like isoform X2</td>\n",
       "      <td>Danio</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>si:ch</td>\n",
       "      <td>si:ch+</td>\n",
       "      <td>Danio_NW_001884452.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Un</td>\n",
       "      <td>NW_003039384.4</td>\n",
       "      <td>161124</td>\n",
       "      <td>165069</td>\n",
       "      <td>+</td>\n",
       "      <td>100136843</td>\n",
       "      <td>zgc:174680</td>\n",
       "      <td>NP_001108034.1</td>\n",
       "      <td>504</td>\n",
       "      <td>uncharacterized protein LOC100136843</td>\n",
       "      <td>Danio</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>zgc:</td>\n",
       "      <td>zgc:+</td>\n",
       "      <td>Danio_NW_003039384.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Un</td>\n",
       "      <td>NW_003039384.4</td>\n",
       "      <td>165289</td>\n",
       "      <td>171203</td>\n",
       "      <td>-</td>\n",
       "      <td>541543</td>\n",
       "      <td>zgc:112970</td>\n",
       "      <td>XP_021326739.1</td>\n",
       "      <td>660</td>\n",
       "      <td>uncharacterized protein LOC541543 isoform X2</td>\n",
       "      <td>Danio</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>zgc:</td>\n",
       "      <td>zgc:-</td>\n",
       "      <td>Danio_NW_003039384.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>Un</td>\n",
       "      <td>NW_003334026.1</td>\n",
       "      <td>60531</td>\n",
       "      <td>70535</td>\n",
       "      <td>-</td>\n",
       "      <td>795664</td>\n",
       "      <td>wu:fa56d06</td>\n",
       "      <td>XP_009296936.1</td>\n",
       "      <td>516</td>\n",
       "      <td>uncharacterized protein LOC795664 isoform X3</td>\n",
       "      <td>Danio</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wu:fa</td>\n",
       "      <td>wu:fa-</td>\n",
       "      <td>Danio_NW_003334026.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18400</th>\n",
       "      <td>MT</td>\n",
       "      <td>NC_023889.1</td>\n",
       "      <td>9918</td>\n",
       "      <td>10214</td>\n",
       "      <td>+</td>\n",
       "      <td>18982988</td>\n",
       "      <td>ND4L</td>\n",
       "      <td>YP_009022039.1</td>\n",
       "      <td>98</td>\n",
       "      <td>NADH dehydrogenase subunit 4L (mitochondrion)</td>\n",
       "      <td>Orca</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND</td>\n",
       "      <td>ND+</td>\n",
       "      <td>Orca_NC_023889.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18401</th>\n",
       "      <td>MT</td>\n",
       "      <td>NC_023889.1</td>\n",
       "      <td>10208</td>\n",
       "      <td>11585</td>\n",
       "      <td>+</td>\n",
       "      <td>18982989</td>\n",
       "      <td>ND4</td>\n",
       "      <td>YP_009022040.1</td>\n",
       "      <td>459</td>\n",
       "      <td>NADH dehydrogenase subunit 4 (mitochondrion)</td>\n",
       "      <td>Orca</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND</td>\n",
       "      <td>ND+</td>\n",
       "      <td>Orca_NC_023889.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18402</th>\n",
       "      <td>MT</td>\n",
       "      <td>NC_023889.1</td>\n",
       "      <td>11787</td>\n",
       "      <td>13607</td>\n",
       "      <td>+</td>\n",
       "      <td>18982990</td>\n",
       "      <td>ND5</td>\n",
       "      <td>YP_009022041.1</td>\n",
       "      <td>606</td>\n",
       "      <td>NADH dehydrogenase subunit 5 (mitochondrion)</td>\n",
       "      <td>Orca</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND</td>\n",
       "      <td>ND+</td>\n",
       "      <td>Orca_NC_023889.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18403</th>\n",
       "      <td>MT</td>\n",
       "      <td>NC_023889.1</td>\n",
       "      <td>13591</td>\n",
       "      <td>14118</td>\n",
       "      <td>-</td>\n",
       "      <td>18982991</td>\n",
       "      <td>ND6</td>\n",
       "      <td>YP_009022042.1</td>\n",
       "      <td>175</td>\n",
       "      <td>NADH dehydrogenase subunit 6 (mitochondrion)</td>\n",
       "      <td>Orca</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND</td>\n",
       "      <td>ND-</td>\n",
       "      <td>Orca_NC_023889.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18404</th>\n",
       "      <td>MT</td>\n",
       "      <td>NC_023889.1</td>\n",
       "      <td>14192</td>\n",
       "      <td>15331</td>\n",
       "      <td>+</td>\n",
       "      <td>18982992</td>\n",
       "      <td>CYTB</td>\n",
       "      <td>YP_009022043.1</td>\n",
       "      <td>379</td>\n",
       "      <td>cytochrome b (mitochondrion)</td>\n",
       "      <td>Orca</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CYTB</td>\n",
       "      <td>CYTB+</td>\n",
       "      <td>Orca_NC_023889.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>467281 rows Ã— 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      #Replicon Name Replicon Accession   Start    Stop Strand     GeneID  \\\n",
       "7                 Un     NW_001878251.4   61071   65540      +     563697   \n",
       "26                Un     NW_001884452.4  167019  169622      +  108183944   \n",
       "37                Un     NW_003039384.4  161124  165069      +  100136843   \n",
       "38                Un     NW_003039384.4  165289  171203      -     541543   \n",
       "52                Un     NW_003334026.1   60531   70535      -     795664   \n",
       "...              ...                ...     ...     ...    ...        ...   \n",
       "18400             MT        NC_023889.1    9918   10214      +   18982988   \n",
       "18401             MT        NC_023889.1   10208   11585      +   18982989   \n",
       "18402             MT        NC_023889.1   11787   13607      +   18982990   \n",
       "18403             MT        NC_023889.1   13591   14118      -   18982991   \n",
       "18404             MT        NC_023889.1   14192   15331      +   18982992   \n",
       "\n",
       "                   Locus Protein product  Length  \\\n",
       "7                 znf995  XP_017209820.1     358   \n",
       "26     si:ch211-198g14.4  XP_021326470.1     266   \n",
       "37            zgc:174680  NP_001108034.1     504   \n",
       "38            zgc:112970  XP_021326739.1     660   \n",
       "52            wu:fa56d06  XP_009296936.1     516   \n",
       "...                  ...             ...     ...   \n",
       "18400               ND4L  YP_009022039.1      98   \n",
       "18401                ND4  YP_009022040.1     459   \n",
       "18402                ND5  YP_009022041.1     606   \n",
       "18403                ND6  YP_009022042.1     175   \n",
       "18404               CYTB  YP_009022043.1     379   \n",
       "\n",
       "                                        Protein name Origin Unnamed: 10  \\\n",
       "7            gastrula zinc finger protein XlCGF8.2DB  Danio         NaN   \n",
       "26             actin-binding protein-like isoform X2  Danio         NaN   \n",
       "37              uncharacterized protein LOC100136843  Danio         NaN   \n",
       "38      uncharacterized protein LOC541543 isoform X2  Danio         NaN   \n",
       "52      uncharacterized protein LOC795664 isoform X3  Danio         NaN   \n",
       "...                                              ...    ...         ...   \n",
       "18400  NADH dehydrogenase subunit 4L (mitochondrion)   Orca         NaN   \n",
       "18401   NADH dehydrogenase subunit 4 (mitochondrion)   Orca         NaN   \n",
       "18402   NADH dehydrogenase subunit 5 (mitochondrion)   Orca         NaN   \n",
       "18403   NADH dehydrogenase subunit 6 (mitochondrion)   Orca         NaN   \n",
       "18404                   cytochrome b (mitochondrion)   Orca         NaN   \n",
       "\n",
       "      Locus tag Gene_non_or    Gene       Origin_Scaffold  \n",
       "7           NaN         znf    znf+  Danio_NW_001878251.4  \n",
       "26          NaN       si:ch  si:ch+  Danio_NW_001884452.4  \n",
       "37          NaN        zgc:   zgc:+  Danio_NW_003039384.4  \n",
       "38          NaN        zgc:   zgc:-  Danio_NW_003039384.4  \n",
       "52          NaN       wu:fa  wu:fa-  Danio_NW_003334026.1  \n",
       "...         ...         ...     ...                   ...  \n",
       "18400       NaN          ND     ND+      Orca_NC_023889.1  \n",
       "18401       NaN          ND     ND+      Orca_NC_023889.1  \n",
       "18402       NaN          ND     ND+      Orca_NC_023889.1  \n",
       "18403       NaN          ND     ND-      Orca_NC_023889.1  \n",
       "18404       NaN        CYTB   CYTB+      Orca_NC_023889.1  \n",
       "\n",
       "[467281 rows x 16 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_genes = df_species[['Gene_non_or', 'Origin_Scaffold']]\n",
    "sca = df_genes['Origin_Scaffold'].unique()\n",
    "sca = pd.DataFrame(sca)\n",
    "sca.rename(columns={0:'Scaffold'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\bertr\\AppData\\Local\\Temp\\ipykernel_12428\\1092919958.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_genes['Next Gene'] = df_genes['Gene_non_or'].shift(-1)\n",
      "C:\\Users\\bertr\\AppData\\Local\\Temp\\ipykernel_12428\\1092919958.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_genes['2 Next Gene'] = df_genes['Gene_non_or'].shift(-2)\n",
      "C:\\Users\\bertr\\AppData\\Local\\Temp\\ipykernel_12428\\1092919958.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_genes['3 Next Gene'] = df_genes['Gene_non_or'].shift(-3)\n",
      "C:\\Users\\bertr\\AppData\\Local\\Temp\\ipykernel_12428\\1092919958.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_genes['4 Next Gene'] = df_genes['Gene_non_or'].shift(-4)\n"
     ]
    }
   ],
   "source": [
    "df_genes['Next Gene'] = df_genes['Gene_non_or'].shift(-1)\n",
    "df_genes['2 Next Gene'] = df_genes['Gene_non_or'].shift(-2)\n",
    "df_genes['3 Next Gene'] = df_genes['Gene_non_or'].shift(-3)\n",
    "df_genes['4 Next Gene'] = df_genes['Gene_non_or'].shift(-4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "ruby"
    }
   },
   "outputs": [],
   "source": [
    "df_genes['3_mers'] = np.where((df_genes['Origin_Scaffold'].shift(-2) == df_genes['Origin_Scaffold']), df_genes['Gene_non_or'] + '_' + df_genes['Gene_non_or'].shift(-1) + '_' + df_genes['2 Next Gene'].shift(-2), np.nan)\n",
    "df_genes['4_mers']= np.where((df_genes['Origin_Scaffold'].shift(-3) == df_genes['Origin_Scaffold']), df_genes['Gene_non_or'] + '_' + df_genes['Next Gene'] + '_' + df_genes['2 Next Gene'] + '_' + df_genes['3 Next Gene'], np.nan)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "vscode": {
     "languageId": "ruby"
    }
   },
   "outputs": [],
   "source": [
    "df_3_mers = df_genes[['Origin_Scaffold', '3_mers']]\n",
    "df_4_mers = df_genes[['Origin_Scaffold', '4_mers']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "vscode": {
     "languageId": "ruby"
    }
   },
   "outputs": [],
   "source": [
    "df_3_mers = df_3_mers.groupby('3_mers', as_index = False).agg(list) \n",
    "df_4_mers = df_4_mers.groupby('4_mers', as_index = False).agg(list) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "vscode": {
     "languageId": "ruby"
    }
   },
   "outputs": [],
   "source": [
    "df_3_mers.to_csv('../Data/Intermediate/3_mers.csv', index=False)\n",
    "df_4_mers.to_csv('../Data/Intermediate/4_mers.csv', index=False)\n",
    "sca.to_csv('../Data/Intermediate/scaffold.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "vscode": {
   "interpreter": {
    "hash": "0e92abe02844fb54062d758795a97b84494055d25efe9d020733c23e9f9966e7"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
